<a href="https://colab.research.google.com/github/joshuajhchoi/ai2learn/blob/master/%ED%8C%8C%EC%9D%B4%EC%8D%AC_%EC%B4%88%EB%B3%B4%EC%9E%90%EB%A5%BC_%EC%9C%84%ED%95%9C_6%EC%8B%9C%EA%B0%84_%EB%A7%8C%EC%97%90_%ED%8C%8C%EC%9D%B4%EC%8D%AC_%EA%B8%B0%EC%B4%88%EB%AC%B8%EB%B2%95%EB%B6%80%ED%84%B0_%EB%A8%B8%EC%8B%A0%EB%9F%AC%EB%8B%9D_%EC%BA%90%EA%B8%80_%EC%9E%85%EB%AC%B8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##목차

####파이썬 초보자를 위한 
####6시간 만에 파이썬 기초문법부터 머신러닝 캐글 입문
1. Colab 사용 기본 (5분)
2. 파이썬 기초 문법 연습 (90분)   
3. NumPy 기본 이해 (20분)
4. Matplotlib 기본 연습 (15분)
5. Seaborn 기본 연습 (30분)
6. Pandas 기본 연습 (30분)
7. ScikitLearn 기본 이해 (50분)
8. Kaggle Titanic 입문하기 (120분)

#PART 1. Colab 사용 기본 (5분)

(1) https://colab.research.google.com/에 가서 회원 가입을 하고 자신의 어카운트를 만듭니다. 

(2) 여기서 사용할 노트는 주피터 노트북 기반이니 모르시는 부분은 주피터 노트북 사이트를 참고하시면 됩니다.https://jupyter.org/ 

(3) 기본적으로 좌측 위에 + 코드를 눌러서 코딩 연습할 공간을 만들고 + 텍스트 부분을 눌러서 설명을 넣을 공간을 만드는 간단한 기능이므로 사용에는 큰 문제는 없을 것입니다.

(4) 이 노트가 6시간 강의 용이라 시간 관계상 자세히 설명할 수는 없으나 메뉴는 각자가 들어가서 쉽게 파악할 수 있으므로 눌러보고 찔러보면 20-30분 내로 모든 것을 파악할 수 있을 것이라 사료 됩니다.

* 간단 설명 영상 

     [Video](https://)


#PART 2. 파이썬 기초 문법 이해하기 (90분)



#Lesson 2– Print Statement


##학습 내용
* 우리가 배울 첫 번째 파이썬 문법은 print 문입니다.
* print문을 통해 사용자는 인쇄하고 결과를 보여줄 수 있습니다.


##문법

         print  ("Hello Python!")

 
##설명

* print (‘문장’)의 형식으로 print 문을 사용합니다.
* 숫자의 경우에는 print (숫자)로 출력합니다.
* 추후 배우겠으나 변수의 값을 출력할 경우에는 print (변수)로 출력합니다.
* Python 2 문법에서는 괄호가 없어서 괄호 없이 인쇄를 해도 되기도 하나 괄호를 넣고 꼭 프린트문을 실행하도록 연습합니다.
* 작은 따옴표 (' ')나 큰 따옴표 (" ") 둘 모두 괜찮으나 둘을 섞여서  
	(' ") 또는 (" ') 형태로 사용되지 않도록 주의합니다. 
* '#' 로 시작하는 라인은 코딩에 영향을 미치지 않으므로 주석이나 설명 등을 넣을 때 사용합니다. 

## 연습

        print ("Hello Python!") 
        print (3)

* 이 두 라인을 에디터에 넣고  Hello Python! 과 3, 이 두개를 출력해 봅니다.


##주의
* 워드 폰트에 따라 따옴표 (＂ ＂) 를 복사하여 Python 에디터에 붙여 넣으면 오류가 발생하기도 합니다. 이 경우 지우고 다시 입력해야 합니다. 




##코딩 부분


In [2]:
# 두 문장을 출력해 봅니다.
print ("Hello Python!") 
print (3)


Hello Python!
3


####Video

[Lesson 2 - Lesson 13까지의 영상](https://youtu.be/zVIQuRShh9U)

##추가 연습 

* 프린트를 할 때 한 줄이 아니라 여러 줄의 문장을 프린트를 하는 경우가 있습니다.  주소 같은 경우가 대표적인 예일 것입니다. 
* 이때는 처음에 따옴표 3개를 넣어서 여러 줄이 시작한다는 것을 컴퓨터에게 알리고 마지막에 따옴표 3개를 넣어서 출력할 내용이 끝났다고 알리면 됩니다.
* 아래 주소를 프린트 문을 사용하여 출력해 봅니다.

        Hong Gildong
        123 Ho, ABC Street,
        Samsung-dong, Gangnam-gu
        Seoul, Korea


In [1]:
print ("""Hong Gildong
123 Ho, ABC Street,
Samsung-dong, Gangnam-gu
Seoul, Korea""")


Hong Gildong
123 Ho, ABC Street,
Samsung-dong, Gangnam-gu
Seoul, Korea


#Lesson 3     -     Method란


##학습 내용
* Method는 기본적으로 내장된 기능(함수)이라고 이해를 하면 됩니다. 
* input()이라는 입력을 위한 method를 우선 배워봅니다.


##문법

        변수 = input()          # 프롬프트가 깜빡 깜빡 나타남
        변수 = input(“질문”)    # 질문 다음에 프롬프트가 깜빡 깜빡 나타남


 
##설명
* 프린트 문으로 출력을 할 수 있다면 그럼 입력은 어떻게 할지 궁금해집니다.  
* ATM기기 앞에서 비밀번호를 입력하는 경우를 보겠습니다.
* 비밀번호를 입력하라는 안내가 나오면 비밀번호 4자리를 입력해 보겠습니다.
* 아래의 경우는 프린트문으로 질문을 출력하고 이어서 아래줄에 깜빡이는 프롬프트에 숫자를 입력하는 방법입니다.

        print ("Please enter your 4 digit code!")
        my_code = input() 

하지만 이 것을 한 줄로 할 수도 있습니다. 이경우 문장의 오른 쪽에 프롬프트가 깜빡입니다.

        my_code = input("Please enter your 4 digit code!") 



## 연습

        print ("Please enter your 4 digit code!")
        my_code = input() 

        my_code = input(“Please enter your 4 digit code!”) 




##주의
* 위의 my_code 같은 것을 변수라 부릅니다. 
* 변수는 소문자로 쓰고 띄어쓰기 대신에 언더바_를 쓰며 다른 것과 이름이 겹치지 않게 써야 합니다.
* 메소드는 함수이나 여기서는 함수와는 다르게 간주한다고 생각하면 됩니다.
* 메소드는 이미 정의되어 내장되어 있는 함수 입니다.
* 따라서 불러올 필요도 없고, 따로 정의할 필요가 없습니다.
* 반면에 이 책에서 말하는 함수는 주로 내가 만드는 함수를 의미 합니다. 



##코딩 부분


In [0]:
print ("Please enter your 4 digit code!")
my_code = input() 

my_code = input("Please enter your 4 digit code!")

Please enter your 4 digit code!
1234
Please enter your 4 digit code!1234


##추가 연습 

* 연습 내용
* "What is your name?" 이란 문장을 프린트하여 이름을 물어보고
* input() 메소드를 통하여 이름을 입력하면 
* Thank you + 입력된 이름을 출력하는 것을 만들어 봅니다.

예)

        What is your name?
        (입력) - 예) Gildong
        Thank you, Gildong


In [0]:
print ("What is your name?") 
your_name = input()
print ("Thank you, " + your_name)

What is your name?
GD
Thank you, GD


* 다음은 input외에 쓰이는 메소드의 일부입니다.

* abs(x) - 숫자의 절댓값을 돌려줍니다. 

In [0]:
print (abs(-3))

3


* all(iterable) - 모든 요소가 참이면 (또는 비어있으면) True 를 돌려줍니다

In [0]:
a = [1, 3, 4, 5]
print (all(a))

True


* bin(x) - 정수를 "0b" 가 앞에 붙은 binary code로 변환합니다. 

In [0]:
a = bin(3)
print (a)

0b11


* bool([x]) - 논리값, 즉 True 또는 False 중 하나를 돌려줍니다.

In [0]:
bl = [0]
print (bool(bl))

True


* chr(i) - 유니코드 코드 포인트가 정수 i 인 문자를 나타내는 문자열을 돌려줍니다. 예를 들어, chr(97) 은 문자열 'a' 를 돌려주고, chr(8364) 는 문자열 '€' 를 돌려줍니다. 


In [0]:
print (chr(97) + chr(98) + chr(99))

abc


# Lesson 4– Strings and Numbers


##학습 내용
* string 은 우리가 배울 첫 번째 유형의 데이터로 알파벳, 기호, 공백 등이 
포함됩니다. 그리고 number 는 숫자형의 데이터입니다.



##문법

        "string"
        3


 
##설명
* string 에는 알파벳, 기호, 공백 및 기타가 포함됩니다.
* string 은 큰 따옴표 (" ") 또는 작은 따옴표 (' ') 에 넣어서 표현을 합니다.
* 예제와 같이 + 를 사용하여 string을 합할 수 있습니다.
* number (숫자)는 우리가 두 번째로 배울 유형의 데이터 입니다.
* 컴퓨터는 string과 number를 다르게 취급합니다.

        예를 들어, 1 은 숫자 1 이지만 "1" 은 문자열 "1" 입니다.
        1 + 1 = 2 이지만  "1" + "1" 은 "11" 입니다
        따라서 2 * 3 은 6 이지만 ‘2’ * 3 은 "222" 입니다.



## 연습

* "I love" 와 "my car" 두 string 사이에 + 를 사용하여 둘을 합하십시오.
* + 를 사용하여 두 개의 숫자 2 와 3 을 더하십시오.
* + 를 사용하여 두 개의 string  "2" 와 "3" 을 합하십시오
* 그리고 결과를 print 하여 봅니다.


##주의
* 단어와 단어를 연결할 때 공백(" ")을 넣어 주는 것이 필요하기도 합니다.
* print 할 때 string 과 number 를 + 를 사용하여 더할 수 없습니다.
* 데이터의 형태가 다르기 때문입니다.





##코딩 부분


In [0]:
print ("I love" + " " + "my car.")

print (2+3)

print ("2" + "3") 

I love my car.
5
23


##추가 연습 

* I was born in 1999을 출력할 때  다음과 같이 출력할 수 없습니다

        print  ("I was born in " + 1999)

* 데이터의 종류가 상이하기 때문입니다.
* 이 경우 1999를 str() method를 사용하여 스트링화 하면 데이터의 종류가 같아져서 프린트 할 수 있습니다.

        a = "I was born in "
        b= str(1999)
        print (a + b)

In [0]:
a = "I was born in "
b= str(1999)
print (a + b)


I was born in 1999


#Lesson 5     -   String Indexing


##학습 내용
* 이 과에서는 string 위치에 접근하는 방법을 배웁니다.
* Indexing이란 표현을 씁니다.


##문법

         스트링변수 [위치]

 
##설명
* HYUNDAI란 단어에는 알파벳이 7자가 있습니다.
* 첫 번째 위치는 1이 아닌 '0'으로 색인되므로 아래 7 개의 문자는 0, 1, 2, 3, 4, 5, 6 의 위치를 갖습니다

        H   Y   U   N   D   A   I
        0   1   2   3   4   5   6



* name이란 변수에 스트링 값을 주고 출력을 print(name[0]) 하면 해당 string의 첫 알파벳이 나옵니다.
* 따라서 HYUNDAI에서 ‘D’의 위치(5번째 위치)를 색인 하려면  [4]를 인덱싱 합니다.

## 연습

* CHEVROLET에서 3번(4번째)의 위치를 indexing 해봅니다.

        maker = "CHEVROLET"
        print (maker[3])




##주의
‘1’이 아닌"0"에서 시작한다는 것을 명심하세요.




##코딩 부분

In [0]:
maker = "CHEVROLET"
print (maker[3])

V


##추가 연습 

* String인 "Korea" 는 알파벳 개수가 5개입니다.
* 0번 부터 4번까지 인덱스가 주어질 수 있습니다.
* 따라서 원하는 것이 r이라면

      "Korea"[2] 라고 하면 r이 나오게 됩니다.


In [0]:
country = "Korea"
print (country[2])

r


#Lesson 6     -   String Methods



##학습 내용
* 세 가지 유형의 대표적인 string 메소드를 배워봅니다. 

        len()   lower()   upper() 


##문법

         	len()  - string의 길이를 보여 줍니다.
        	lower() - string을 모두 소문자로 변환합니다.
        	upper() - string을 모두 대문자로 변환 합니다.


 
##설명
* import 나 새로 정의를 안 해도 쓸 수 있는 함수를 메소드라고 합니다.

다음의 예를 보며 위의 메소드를 이해 하겠습니다.

        maker란 변수에  'Mercedes'란 값을 줍니다.
        maker = 'Mercedes'

* 여기서 주어진 스트링을 길이를 구하려면 

      len (maker)를 하면 됩니다. 
      그러면 8이 답으로 나옵니다.

*숫자의 길이를 구할 수는 없습니다.
* len(12345)는 에러를 발생합니다. 
* 긴 숫자는 긴 것이 아니라 큰 것이기 때문입니다.
* 반면에 str() 메소드를 써서 len(str(12345))를 프린트하면 에러가 발생하지 않고 5가 나옵니다.


## 연습
* 다음을 해봅니다. 


        maker = "CHEVROLET"
        print (len(maker))
        print (len("Hong Gildong")
        a = 12345
        print (len(str(a)))




##주의
* 빈칸도 길이에 들어 갑니다. 




##코딩 부분

In [0]:
maker = "CHEVROLET"
print (len(maker))
print (len("Hong Gildong"))
a = 12345
print (len(str(a)))

9
12
5


##추가 연습 

* 메서드 lower()와 upper()는 기존에 사용했던 str(123) 메서드와 달리 다음과 같이 표현하며 사용합니다.


        maker.lower()
        maker.upper()


* lower()는 string을 모두 소문자로 만들어 주고 upper()는 string을 모두 대문자로 만들어 줍니다.
* 이와 같이 표현하는 방법을 “Dot Notation”이라고 부릅니다.
* 구체적인 내용은 함수를 배우면서 더욱 익숙해질 것입니다.

In [0]:
maker = "CHEVROLET"
print (maker.lower())
print (maker.upper())

chevrolet
CHEVROLET


#Lesson 07    -    Number Methods


##학습 내용
* 두 가지 유형의 숫자 데이터가 있습니다 
* integer(정수)와 float(실수) 및 그와 관련된 메소드를 보겠습니다.


##문법

        int(숫자)   # 정수로 표현된다
        float(숫자) # 실수로 표현된다
        str(숫자)   # 보이기는 숫자로 보이지만 문자로 인식된다.



 
##설명
* integer는 소수점 이하의 숫자를 나타낼 수 없습니다. 
* float은 소수점 이하의 숫자를 나타낼 수 있습니다. 
 
        int_sample = 1
        int_test = 2
        int_minus = -3
        float_sample = 1.0
        float_test = 2.
        float_minus = -3.5

* 그 외에, 숫자의 단위 수를 높여 0을 늘릴 때 e는 유용합니다.

예를 들어 

        float_e = 3.5e3는 3.5 X 1,000을 의미합니다.

* %는 나머지를 나타냅니다.   3%10 = 1 



## 연습


* float 3.5를 int(3.5)로 하여 3이 나오는지 확인해 봅니다.
* int 3을 float(3)으로 하여 3.0이 나오는지 확인해 봅니다.
* int(3.5) + float(3)의 답이 얼마가 나오는지 확인해 봅니다.
* str(3) + str(3.5)를 더해 봅니다.


##주의
* 참고로 Input() 메소드로 입력되는 number는 string으로 인식됩니다. 
* 따라서 입력된 숫자를 숫자로 사용하려면 int(입력된 숫자)로 바꾸어 주어야 합니다. 

* 아래의 예를 봅니다.

        input_number = input()
        my_number = int(input_number)
        # 이렇게 바꾸면 my_number의 값이 string이 아니라 integer가 됨 
        # 또는 처음부터 아래와 같이 integer로 인식하게 하는 것도 가능
        a_number = int(input())




##코딩 부분

In [0]:
test1 = int(3.5)
test2 = float(3)
test3 = test1 + test2
test4 = str(3) +str(3.5)

print (test1)
print (test2)
print (test3)
print (test4)

3
3.0
6.0
33.5


##추가 연습 

* 변수 gas_liter에 integer 5.5의 값을 줍니다. (구매할 휘발유의 리터량)
* price_per_liter 변수를 만들고 3의 값을 줍니다. (휘발유 리터 당 3달러)
* 변수 sum_price를 생성하고 구매할 gas의 총 가격을 계산하여 할당합니다. (총 구매 액수)
* print 문을 사용하여 sum_price 변수의 integer값을 인쇄합니다.
* 같은 값을 float 값으로 인쇄합니다.


In [0]:
gas_liter =5.5
price_per_liter = 3
sum_price = gas_liter * price_per_liter
print (int(sum_price))
print (float(sum_price))

16
16.5


####Extra Practice


(1) 
* 변수 apple_count 에 정수 10 을 대입하세요.
* 변수 price_apple 를 만들고, 3000을 대입하세요.
* 변수 sum_price 를 만들고, 구매 할 사과의 총 가격을 계산하여 대입하세요.
* 변수 sum_price 를 print 문을 사용하여 출력하세요.

In [0]:
apple_count=10
price_apple = 3000
sum_price = apple_count * price_apple
print (sum_price)

30000


2) 
* (1)의 결과는 정수로 나온다.
* 하지만 사과를 반쪽을 살 수 있다고 가정해봅니다.
* 변수 apple_count 에 실수 5.5를 대입하세요.
* 변수 price_apple 를 만들고, 3000을 대입하세요.
* 변수 sum_price 를 만들고, 구매 할 사과의 총 가격을 계산하여 대입하세요.
* 변수 sum_price 를 print 문을 사용하여 출력하세요.


In [0]:
apple_count=5.5
price_apple = 3000
sum_price = apple_count * price_apple
print (sum_price)

16500.0


#Lesson 08    -    Boolean Operators


##학습 내용
* 데이터를 True 및 False로 표현해야 하는 경우가 있습니다.
* 이는 string, numbers에 이어서 세번째로 배울 데이터 타입입니다.


##문법

         a >  b     a는 b보다 큼                   
         a < b      a는 b보다 작음
         a == b     a와 b은 같다                   
         a != b    a와 b는 같지 않음
         a <= b     a는 b보다 작거나 같음      
         a >= b     a는 b보다 크거나 같음


 
##설명
* Boolean은 True / False로 나오나 숫자 값을 가집니다.
* True 는 결과값이 참일 때 반환된다. (True의 값은 1입니다.)
* False 는 결과값이 거짓일 때 반환된다. (False 값은 0입니다.)
* 다음 연산자를 알아야합니다. (아래는 True가 되는 내용입니다.)

    	3 > 2       (3은 2보다 큼)
    	1 <2        (1이 2보다 작음)
    	1 == 1      (1과 1은 같다 )
    	1 ! = 2     (1과 2가 같지 않음)
    	2 <= 2      (2는 2보다 작거나 같음)
    	2 >= 2      (2는 2보다 크거나 같음)


## 연습


* 아래와 같이 정하여 각각의 연산자를 비교하고 그 값을 확인해 봅니다.

        testnumber1 = 1
        testnumber2 = 2
        testnumber3 = 3



##주의
* 조건문에는 = 가 아니라 ==를 사용한다는 것을 기억하세요



##코딩 부분

In [0]:
testnumber1 = 1
testnumber2 = 2
testnumber3 = 3

a = testnumber3 > 2
b = testnumber1< 2 
c = testnumber1 == 1 
d = testnumber1 != 2
e = testnumber2 <= 2 
f = testnumber2 >= 2

print (a)
print (b)
print (c)
print (d)
print (e)
print (f)


True
True
True
True
True
True


      #숫자를 바꾸어서 False가 되는 경우도 연습합니다.

##추가 연습 

* 아직 배우지 않은 것이지만 조건문의 맛의 잠깐 보겠습니다.
* 아래를 실행을 해보면 이해가 갈 것입니다.

 
        x = 5 

        if x > 5:
            print ( "It is over the limit." )
        else:
            print ( "It is under the limit." )


* if-else는 Task 12에서 배우게 될 것이고 여기서는 불리안의 쓰임만 봅니다.


        X = 6으로 고쳐서 다시 한 번 해봅니다.


In [0]:
x = 5 
if x > 5:
   print ("It is over the limit.")
else:
   print ("It is under the limit.")

It is under the limit.


    # 6으로 고쳐 넣으면 아래와 같이 출력이 됩니다.
    It is over the limit.

#Lesson 09   -       Variables


##학습 내용
* 앞에서 우리는 변수(variable)를 접했습니다. 
* 수학에서는 Y = X + 2와 같은 식을 볼 수 있으며 Y와 X 값은 변하지만 2는 변하지 않습니다. 그래서 X,Y는 변수입니다. 파이썬에서의 변수의 쓰임새를 살펴보겠습니다.


##문법

        variable = "string"     # 문자형 변수
        variable = 5            # 숫자형 변수
        variable = input()      # 함수형 변수


 
##설명
* 수학에서는 Y = X + 2와 같은 것을 볼 수 있으며 Y와 X 값은 변하지만 2는 변하지 않습니다. 그래서 변수와 상수라는 표현을 씁니다.
* 우리는 Y와 X를 변수라고 부릅니다.
* 변수(variable)에는 값이 있습니다.
* 이러한 값은 ＂string’, number (integer 또는 float) 또는 True / False 값 등이 있다고 배웠습니다.
* 변수 = ‘string’과 변수 = 숫자 두 가지를 연습해 봅니다.


## 연습



* 문자형 변수의 조합을 살펴 보겠습니다. Peter가 차를 사려고 합니다. 결정한 것은 없는데 Josh가 새 차 및 중고차의 두 경우에 대비하여 3가지 답을 생각해 놓았는데 이를 섞어서 4가지 답을 만들어 봅니다.
* 문자형 변수를 결합해 봅니다.


        situation1 = "It is a new car!"   
        situation2 = "It is a used car."
        comment1 = "Too expensive!!"
        comment2 = "It is a good deal!"
        comment3 = "It can give you headaches!"



##코딩 부분

In [0]:
situation1 = "It is a new car!"   
situation2 = "It is a used car."
comment1 = "Too expensive!!"
comment2 = "It is a good deal!"
comment3 = "It can give you headaches!"


answer1 = situation1 + " " + comment1
answer2 = situation1 + " " + comment2
answer3 = situation2 + " " + comment3
answer4 = situation2 + " " + comment2

print (answer1)
print (answer2)
print (answer3) 
print (answer4)


It is a new car! Too expensive!!
It is a new car! It is a good deal!
It is a used car. It can give you headaches!
It is a used car. It is a good deal!


##추가 연습 

* 스트링과 숫자를 합하는 연습을 간단히 해봅니다.
* 아래 코드와 설명을 보고 이해를 한 번 해봅니다.


        # 우리 강아지 이름은 코코예요
        my_dog_name = ‘Coco’ 	
        
        #코코는 한 주에 4번 산책해요
        walk_per_week = 4   	
        
        # 한 번 산책 할 때 마다 5곳에 들려요
        point_visit = 5  				
        
        # 총 들려야 하는 곳은 130 곳이예요
        points_to_visit = 130  		
        
        # 몇 주 걸릴 것인가?
        weeks_to_take = float(points_to_visit) / (walk_per_week * point_visit )  	


In [0]:
my_dog_name = "Coco"
walk_per_week = 4
point_visit = 5
points_to_visit = 130
weeks_to_take = float(points_to_visit) / (walk_per_week * point_visit )
statement = my_dog_name + " needs " + str(weeks_to_take) + " weeks to visit those " + str(points_to_visit) + " places."
print (statement)


Coco needs 6.5 weeks to visit those 130 places.


#Lesson 10    -    Updating Variables


##학습 내용
* 변수는 업데이트하여 변경할 수 있습니다. 
* 마지막으로 대입되는 것이 최종 값이 됩니다.


##문법

        x = 1
        x = x + 1 
        print (x)  # x = 2로 갱신된다.

 
##설명
* 변수는 업데이트하여 변경할 수 있습니다.
* 예를 들어 x = 1이라 정의하였는데 아래에 x = x + 1이라 재 정의하면 x는 2가 됩니다.
* 이와 같이 프로그램에서는 가장 아래 것 (또는 가장 최신 것)이 값을 갱신합니다.
* 마찬가지로 기름통에 10리터가 있는데
* 변수 = 남은 기름 양으로 한다면
* 쓸 때 마다 그 값은 줄어드는 것이 계속 업데이트 되어져야 할 것입니다.
* 변수는 숫자형 변수였다가 스트링형 변수로도 변할 수 있습니다.

## 연습


* 변수 gas_tank에 10의 값을 대입합니다. (가스 통에 10리터 있다고 가정)
* remaining gasoline(사용 후 남은 양)을 인쇄합니다.
* 변수 gas_used (사용된 휘발유)에 5를 대입합니다.
* 변수 gas_tank에서 gas_used를 빼면 탱크에 남아있는 휘발유 값이 나오겠죠.
* 남은 양을 인쇄합니다.



##주의
* 갱신된다는 것은 변수의 값이 바뀐다는 것을 의미합니다. 



##코딩 부분

In [0]:
gas_tank = 10

print ( "remaining gasoline" )

gas_used = 5

gas_tank = gas_tank - gas_used

print (gas_tank)



remaining gasoline
5


##추가 연습 

* 변수가 숫자였다가 스트링으로 업데이트 될 수도 있습니다.
* 변수에 숫자 값을 줍니다.
* 그리고 이를 프린트 합니다.
* 같은 변수 이름에 스트링 값을 줍니다.
* 이를 프린트 해 봅니다.


In [0]:
my_var = 1
print (my_var)
my_var = "Hello"
print (my_var)

1
Hello


#TASK 11      - String Format


##학습 내용
* string과 number를 섞어서 출력할 때 우리는 숫자를 스트링화 시켜서 문장을 만들었습니다. 
* 이제 이를 삽입하여 쓸 수 있는 format 법을 배웁니다.


##문법
* 문장에 {}를 넣고 인수의 위치를 지정해준 후 .format()에 데이터를 넣습니다.

        print  ("I am {0}, and I love {1} ".format (" Peter "," BMW")
        #  I am Peter, and I love BMW


 
##설명
* Python 3.x 버전에서는 string 메소드 format() 을 사용합니다.
* {0}, {1}, {2}, {3}과 같은 인수의 위치를 제공합니다.
* 연습을 하다 보면 Python 2.x 버전에서 유사하게 사용한 것을 보실 수 있는데, 이는 추가 연습에서 다시 한 번 보시겠습니다. 


## 연습


* 아래와 같이 자신의 이름을 넣고 연습해 봅니다.


        name = '홍길동'
        age  = 23 
        
        print ("제 이름은  {} 입니다.".format(name))
        print ("저는  {}세 입니다.".format(age))


##주의
* 위의 경우는 {}이 한 개이나 여러 개의 경우 중괄호 {} 안에 0,1 등으로 정해줘야 합니다.




##코딩 부분

In [0]:
name = "홍길동"
age  = 23 

print ("제 이름은  {} 입니다.".format(name))
print ("저는  {}세 입니다." .format(age))

제 이름은  홍길동 입니다.
저는  23세 입니다.


##추가 연습 

* 다른 방법으로 포매팅을 해보겠습니다.
* 아직 배우지 않은 것이지만 리스트의 개념, 임포트의 개념, math를 import해서 max, min, mean의 개념을 사용해 봅니다.

			# 리스트 age
			age = [22, 35, 32, 7, 67, 34, 21]
			# math 란 모듈을 import 해옵니다.
			import math
			# 여기서 여기서 max (최대값 찾기) min (최소값 찾기) mean (평균 값 찾기) 메소드를 사용하여 이를 포매팅한 문장을 프린트 해봅니다. 두 방법으로 해봅니다.
			# 평균은 파이썬 기본이 아니라 numpy 함수라 numpy를 import 합니다.
			import numpy as np

			
			print('Oldest person is',max(age),'years old.')					
			print('Youngest person is',min(age),'years old.')					
			print('Average age is', np.mean(age),'years old.')

			print('Oldest person is {} years old.'.format(max(age)))					
			print('Youngest person is {} years old.'.format(min(age)))					
			print('Average age is {} years old.'.format(np.mean(age)))


* 위와 같이 좀 더 추가적인 기능을 사용하려면 import해야 하는데 아래 추후 레슨에서 배우시도록 하고 여기서는 맛만 보았습니다.

In [11]:
age = [22, 35, 32, 7, 67, 34, 21]
import math
import numpy as np

print('Oldest person is',max(age),'years old.')					
print('Youngest person is',min(age),'years old.')					
print('Average age is', np.mean(age),'years old.')

print('Oldest person is {} years old.'.format(max(age)))					
print('Youngest person is {} years old.'.format(min(age)))					
print('Average age is {} years old.'.format(np.mean(age)))


Oldest person is 67 years old.
Youngest person is 7 years old.
Average age is 31.142857142857142 years old.
Oldest person is 67 years old.
Youngest person is 7 years old.
Average age is 31.142857142857142 years old.


#  Lesson 12    -    If문과 Else문



##학습 내용
* if 문은 조건이 옳다면 다음에 오는 들여 쓰기 한 코드에 따라 무언가를 하라고 지정하는 것입니다.
* else문 if 조건이 일치하지 않을 때 다른 작업을 수행하라고 하는 문법입니다.


##문법

* if - else의 문법은 if문과 그에 종속된 body 그리고 else문과 그에 종속된 body로 구속되어 있다.

        if condition:
            body
        else:
            body


 
##설명
* if 문은 조건이 옳다면 다음에 오는 들여 쓰기 한 코드에 따라 무언가를 하라고 지정하는 것입니다.
* else문은 if 조건이 일치하지 않을 때 다른 작업을 수행하라고 하는 문법입니다.
* 예를 들어 자동차가 일정속도 이상을 달릴 경우(if) 경고를 주고
* 아닌 경우 (else) 속도를 내라고 하는 명령을 프린트 할 수 있겠습니다.
* 또는 그 반대나 다른 조건들을 만들 수도 있습니다.

## 연습

    "How fast are you going?"의 질문을 프린트 합니다.
            (깜빡) --> 속도 입력
    속도가 50 이하인 경우 
            "Go faster."를 프린트하고
    아닌 경우 
            "Slow down."을 프린트 합니다.



##주의
* else문을 없앨 경우, else문이 없다고 if문이 프린트가 안되는 것은 아니나 그 경우 자기 조건에 맞지 않는 것은 프린트를 못 합니다.



##코딩 부분

In [0]:
my_speed = input("How fast are you going?")
int_speed = int(my_speed)

if int_speed <= 50:
   print ("Go faster")

else:
  print ("Slow down.")


##추가 연습 

* 운전을 하는데
            앞에 신호등에 빨간 불이면  1을 입력 
                     "Stop the car."를 프린트하고
            아닌 경우 2를 입력
                     "Go forward."을 프린트 합니다.

In [0]:
print ("If the traffic light is red, enter 1. if the traffic light is blue or yellow, enter 2.")
t_code = int(input())

if t_code == 1:
  print ("Stop the car.")
else :
    print ("Go forward.")

If the traffic light is red, enter 1. if the traffic light is blue or yellow, enter 2.
2
Go forward.


#TASK 13    -    If문, Elif문 & Else문


##학습 내용
* if–else가 이 것 아니면 나머지 경우라면 if-elif-else는 다양한 옵션을 줍니다.  
* elif은 여러 개가 가능하므로 다양한 조건을 표현할 수 있습니다.



##문법

* if문 - elif문 - else문은 아래와 같은 구조를 가집니다.

            if condition:
                body
            elif condition:
                body
            else
                body


 
##설명
* 앞 과에서 if-else 를 연습하였습니다.
* 이 과에서는 같은 예제들을 if –elif—else로 해보겠습니다.
* 신호등을 보고 운전하고 있다고 가정 해 보겠습니다.
    
    	if 푸른색:
       		print ("지나가세요")
    	elif 빨간색:
       		print ("정지하세요.")
    	else:
      		print ("서행하세요.")

 

## 연습
* 앞 과에서 한 연습을 살짝 바꾸어 봅니다.

      "How fast are you going?"의 질문을 프린트 합니다.
      (깜빡) --> 속도 입력
      속도가 50 이하인 경우   "You can go faster."를 프린트하고
      속도가 80이상인 경우 "Slow down."을 프린트 합니다.
      둘 다 아닌 경우 (50초과 80미만)는 "Okay"를 프린트합니다.


##주의
* elif은 하나 이상, 여러 개 사용이 가능 합니다.



##코딩 부분

In [0]:
my_speed = input("How fast are you going?")
int_speed = int(my_speed)

if int_speed <= 50:
   print ("Go faster")
elif int_speed >= 80:
   print ("Slow down.")
else:
  print ("Okay.")


##추가 연습 

* 앞 과에서는 신호등이 빨간 색이면 멈추고 아니면 간다는 연습을 했는데 이번엔 빨간색이면 멈추고, 노란색이면 속도를 줄이고, 아니면 (녹색) 앞으로 간다를 해보겠습니다. 

In [0]:
print ("If the traffic light is red, enter 1, if the traffic light is yellow, enter 2, and if the traffic light is green, enter 3.")
t_code = int(input())

if t_code == 1:
  print ("Stop the car.")
elif t_code == 2:
  print ("Slow down.")
else :
    print ("Go forward.")

If the traffic light is red, enter 1, if the traffic light is yellow, enter 2, and if the traffic light is green, enter 3.
2
Slow down.


#  TASK 14       -      List 



##학습 내용
* 리스트는 대괄호 ‘[ ]’로 표현되며 데이터를 하나의 변수로 관리할 때 사용됩니다. 
* 변경 가능하고, 중복도 가능하여 일반적인 목록으로 사용된다.  


##문법
* 리스트는 아래와 같은 구조를 가집니다.

         my_list = [1, 2, 3, "a", "b", "c"]

 
##설명
* "list" 는 말 그대로 목록입니다.
* 집합과 비슷한데 연속적으로 놓여있는 리스트의 아이템들을 포함합니다.
* 대괄호 [Square Bracket]로 표현되며 데이터를 하나의 변수로 관리할 때 사용됩니다.
* 리스트는 또한 다른 리스트를 자신의 아이템으로 가질 수 있습니다.

  즉 리스트 안에 리스트가 올 수 있다는 것입니다.

* 리스트의 아이템은 교체가 가능하여 교체가 안 되는 tupple과 비교되며, 아이템 중복이 가능하여 중복이 안 되는 set과 구별됩니다. 
 

## 연습
* 1,2,3 의 단순한 숫자를 아이템으로 하는 리스트를 만들어 봅니다.

        my_list1 = [1, 2, 3] 


* 그리고 string으로 된 아이템들을 만들어 봅니다.

        my_list2 = ["KIA", "Chevrolet", "Ford"]

* 이 들을 프린트 해 봅니다.


##주의
* 리스트는 리스트 안에 숫자, 문자를 같이 넣을 수 있고 다른 리스트를 넣을 수 있습니다.



##코딩 부분

In [0]:
my_list1 = [1, 2, 3]
my_list2 = ["KIA", "Chevrolet", "Ford"]
print (my_list1)
print (my_list2)

[1, 2, 3]
['KIA', 'Chevrolet', 'Ford']


####Video
[Lesson 14 - Lesson 27까지의 영상](https://youtu.be/zVIQuRShh9U)

##추가 연습 

        my_list3 = [1, 2, 3, ["KIA", "Chevrolet", "Ford"]]


* 이번에는 1,2,3에 이어서 4번째 아이템을 string으로 된 리스트를 아이템으로 둔 리스트를 봅니다.
* 이 경우 리스트는 아이템을 4개로 표현할 지 6개로 표현할지 len()으로 검사 해봅니다.
* 4번째 요소가 리스트로 표현될까? 한 번 살펴봅니다.


In [0]:
my_list3 = [1, 2, 3, ["KIA", "Chevrolet", "Ford"] ]
print (len(my_list3))
print (my_list3[0])
print (my_list3[3])



4
1
['KIA', 'Chevrolet', 'Ford']


####Extra Practice

(1)   슬라이스 인덱싱

        my_list = [1,2,3,4,5]
        print my_list[0:2] #첫째부터 셋째 이전까지 출력 (즉 [1] [2] 위치 출력)
        print my_list[1:] # 둘째부터 끝까지 출력 (즉, [1] 위치부터 출력) 
        # 이를 응용하여, 잘라서 어디까지 만들 수 있나 봅니다.


In [0]:
my_list = [1,2,3,4,5]
print (my_list[0:2]) #첫째부터 3째 이전까 출력
print (my_list[1:]) # 둘째부터 끝까지 출력

[1, 2]
[2, 3, 4, 5]


(2)   Replacing

        my_list = [1,2,3,4,5]
        my_list[1] = 6
        print my_list
        #여기서 둘째 위치가 6으로 바뀌는지 확인해 봅니다.


In [0]:
my_list = [1,2,3,4,5]
my_list[1] = 6
print (my_list)


[1, 6, 3, 4, 5]


(3)  Adding


        # 리스트를 더해 봅니다.
        my_list1 = [1,2,3,4,5]  
        my_list2 = [6,7,8,9]
        my_list = my_list1 + my_list2
        print my_list


In [0]:
my_list1 = [1,2,3,4,5]
my_list2 = [6,7,8,9]
my_list = my_list1 + my_list2
print (my_list)

[1, 2, 3, 4, 5, 6, 7, 8, 9]


(4) Multiplying

In [0]:
my_list1 = [1,2,3,4,5]
my_list = my_list1 *2
print (my_list)


[1, 2, 3, 4, 5, 1, 2, 3, 4, 5]


#TASK 15     -   List Methods 


##학습 내용
* 리스트의 메소드 중 append(), count(), index(), insert(), pop(), extract()에  대해서 학습해 봅니다.

##문법
* 아래와 같은 메소드를 공부해 봅니다.

        listname.append(item)  
        listname.count(item)
        listname.index(item)
        listname.insert(위치, item)
        listname.pop(위치)


 
##설명
* list.append()     # 리스트에 아이템을 추가한다. append를 하면 추가된 아이템이 끝에 추가된다.
* list.count()      # 특정 아이템이 몇 개가 있는지를 알려준다.
* list.index()      # 특정 아이템이 어느 위치에 있는지 알려준다.
* list.insert()     # 리스트에 아이템을 넣어준다. insert를 3번째 위치에 했다면 insert된 아이템은 그 위치에 들어가고 본래 그 위치에 있던 아이템 및 그 뒤 아이템들은 뒤로 밀린다.
* list.pop()        # 리스트의 특정 아이템을 보여주고 사라지게 한다.


## 연습
* 빈 리스트 my_list4를 만들어 봅니다.
* 여기에 ＂Ford’ 를 append로 넣어 봅니다.
* ‘Ford’ 가 리스트에 몇 개가 있는지 확인합니다.
* ‘Ford’가 어디에 있는지 확인합니다.
* 이어서 ＂BMW’를 리스트에 넣어 봅니다.
* pop method를 사용하여 제일 앞에 있는 값이 보여지게 합니다.
* 이어서 insert를 사용하여 ＂Chevrolet’ 을 두 번째 위치에 넣어봅니다.
* my_list4를 출력해봅니다.



##주의
pop()을 사용하면 해당 아이템이 보여진 후 리스트내에서 삭제되므로 주의해야 합니다.




##코딩 부분

In [0]:
my_list4 = []
my_list4.append("Ford")
print (my_list4)
print (my_list4.count("Ford"))
print (my_list4.index("Ford"))
my_list4.append("BMW")
print (my_list4)

print (my_list4.pop(0))
my_list4.insert(1,"Chevrolet")
print (my_list4)



##결과 콘솔

        ["Ford"]            # 이를 통해서 빈 리스트에＂Ford’가 추가 되었음을 확인함 
        1                   # 그래서 아이템이 한 개 임을 확인함
        0                   # 이는 ＂ Ford’가 첫 위치에 있음을 확인함
        ["Ford", "BMW"]     # 이는 ＂BMW’가 추가 되었음을 확인함
        Ford                # 첫 아이템이 ＂Ford’ 나타나게 함
        ["BMW", "Chevrolet"]# 이는 위에서 ＂Ford’가 pop 한 다음에 사라지고
                            1의 위치 (2번째 위치에) 'Chevrolet'이 추가되고
                            BMW는 두 번째 위치에서 ＂Ford’가 사라진 후 
                            첫 위치로 갔음을 알게 해 준다.

##추가 연습 

* 정규 표현식
* 정규 표현식(Regular Expressions)이란 복잡한 문자열을 처리할 때 사용하는 기법이다. 예를 들어 주민등록번호 991224 - 1234567 에서 뒤를 991224 -1****** 로 바꾼다든지 하는데 도움을 준다. 
* impore re 를 하여 도움을 받거나 pandas 등의 모듈을 import 해야 한다.
* 아래 예제를 해본다.
* 리스트 name = ['Mr. James','Ms. Johnson'. 'Dr. Hammels', 'Rev. Kim', 'Mr. Gray', 'Mrs. Daisy']에서 Mr. 같은 Salutation만 추출해 salute란 리스트로 만들어 봅니다. 

      name = ['Mr. James','Ms. Johnson'. 'Dr. Hammels', 'Rev. Kim', 'Mr. Gray', 'Mrs. Daisy']
      salute = name.str.extract('([A-Za-z]+)\.') 

* 위는 주어진 리스트에서 .으로 나눈 후 .의 앞 부분[0] (뒷 부분일 경우 [1])을 가지고 새로운 리스트를 만들라는 의미이다.
* 추후 regex란 메소드를 가지고 만들 수도 있으나 이는 나중에 다시 배워 보기로 한다.

In [23]:
name = ['Mr. James','Ms. Johnson', 'Dr. Hammels', 'Rev. Kim', 'Mr. Gray', 'Mrs. Daisy']
salute = [x.split('.')[0] for x in name]
print (salute)


['Mr', 'Ms', 'Dr', 'Rev', 'Mr', 'Mrs']


####Extra Practice

(1) append

In [0]:
list1 = [1,2,3,4]
list1.append("lion")
print (list1)


[1, 2, 3, 4, 'lion']


(2)  count


In [0]:
list2 = [1,2,3,4,1,2,3,4]
print (list2.count(2))

2


(3)  index

In [0]:

list3 = [4,5,6,7,8,9]
print (list3.index(7))

3


(4) insert

In [0]:
list4 = [9,8,7,6,5,4]
list4.insert(3, 100)
print (list4)

[9, 8, 7, 100, 6, 5, 4]


(5)  pop

In [0]:
list5 = [3,4,5,6]
list5.pop(1)
print (list5)

[3, 5, 6]


#Lesson 16       -   Dictionary 



##학습 내용
* "list"가 하나의 variable로 관리되는 목록이라면 "dictionary" 는 사전이나 명함에 좀 더 가깝다고 보면 되겠습니다. Dictionary는 대괄호{ } 로 표현됩니다. 

##문법
* 딕쇼너리는 아래와 같은 구조를 가집니다.

         mydict = {key: value, key: value, key; value}


 
##설명
* 아래의 예를 보면 이름, 전화번호, 회사명이 나와 있습니다. 이러한 것을 관리하는 목록이라 생각하면 됩니다. 

        Name : Joshua
        Phone : 55512345678
        Company: ai2learn

* 여기서 "Name" 은 Key라 부르고 "Joshua" 는 value라 부릅니다.
* 항목과 그에 대한 값이 있는 형태입니다.


## 연습
* bizcard란 이름의 빈 dictionary를 만들어 보겠습니다. 

        bizcard = {}


* 그리고 아래와 같은 요소들을 넣어 봅니다.

        bizcard['name'] = 'Joshua'
        bizcard['phone'] = '55512345678'
        bizcard['company'] = 'ai2learn'

*이어서 bizcard를 출력하고 한 요소인 phone 만 출력합니다.



##주의
* Dictionary도 key나 value의 변동은 허용하나 key의 겹침은 허용 안됨




##코딩 부분

In [0]:
bizcard = {}
bizcard["name"] = "Joshua"
bizcard["phone"] = "55512345678"
bizcard["company"] = "ai2learn"

print (bizcard)
print (bizcard["phone"])


{'name': 'Joshua', 'phone': '55512345678', 'company': 'ai2learn'}
55512345678


##추가 연습 

* 다른 이름을 key와 value를 넣어서 한 번 더 해봅니다.

        이름 : Ironman
        전화번호:  01012345678   
        회사명 : Shields

* 이제 해당 딕쇼너리에서 전화 번호를 찾아내 봅니다.


In [0]:
bizcard = {}
bizcard['이름'] = 'Ironman'
bizcard['전화번호'] = '012345678'
bizcard['회사명'] = 'shields'

print (bizcard)
print (bizcard['전화번호'])

{'이름': 'Ironman', '전화번호': '012345678', '회사명': 'shields'}
012345678


#Lesson 17     -   Dictionary Methods


##학습 내용
* Dictionary의 메소드 중 items(), keys(), values()에  대해서 학습해 봅니다.

##문법
* 다음과 같은 메소드를 공부해 봅니다.

	    dictionaryname.items() 
	    dictionaryname.keys() 
	    dictionaryname.values() 


 
##설명
*   bizcard.items()         # 이 메써드는 dictionary 아이템들을 list 형식으로 변환해 준다.
*   bizcard.keys()          # 이 메써드는 dictionary의 key를  list 형식으로 변환해 준다.
*   bizcard.values()    # 이 메써드는 dictionary의 value를  list 형식으로 변환해 준다.


## 연습
* 배운 것을 아래 key와 value로 연습해봅니다
* 
        model : "Camaro"
        maker : "Chevrolet"
        year: "2020"


* item을 프린트 합니다.
* key를 프린트 합니다.
* value를 프린트 합니다.



##주의
* Key와 value값에서 스트링에 따옴표를 하는 것을 잊지 마세요.



##코딩 부분


In [0]:
mydict = { 'model' : 'Camaro', 'maker': 'Chevrolet', 'year': 2020}

print (mydict.items())
print (mydict.keys()) 
print (mydict.values())


dict_items([('model', 'Camaro'), ('maker', 'Chevrolet'), ('year', 2020)])
dict_keys(['model', 'maker', 'year'])
dict_values(['Camaro', 'Chevrolet', 2020])


##추가 연습 

* 추가적으로 다음과 같은 메소드들이 있습니다 

        clear()     # dictionary의 모든 요소들을 없앤다.
        copy()      # dictionary의 복제된 값을 반환한다
        get()       # 특정 key의 value값을 반환한다

* 본 연습의 예를 보고 다음을 해봅니다. 


        mydict = { 'model' : 'Camaro', 'maker': 'Chevrolet', 'year': 2020}
        newdict = mydict.copy()  
        print (newdict)
        print (mydict.get('model'))
        print (mydict.clear())


In [0]:
mydict = { 'model' : 'Camaro', 'maker': 'Chevrolet', 'year': 2020}
newdict = mydict.copy()  
print (newdict)
print (mydict.get('model'))
print (mydict.clear())

{'model': 'Camaro', 'maker': 'Chevrolet', 'year': 2020}
Camaro
None


#Lesson 18       -     Tupple


##학습 내용
* tupple이란 변경되지 않은 데이터의 집합으로 변경되지 못 한다는 특성 때문에 메소드가 별로 없습니다. tupple은 소괄호 ( )로 표현됩니다. 


##문법
* 튜플은 아래와 같은 형태를 가집니다.

         mytupple = (item1, item2, item 3)


 
##설명
* 그럼 왜 tupple을 사용할까요? 다음과 같은 이유 때문입니다.
* tupple은 저장 공간을 적게 쓴다.
* tupple은 우리가 우연하게 또는 실수로 데이터를 바꾸어서는 안 되는 곳에서 사용된다.
* 더하여 합할 수 있고 곱하여 반복이 가능하다. 

        a = ('Chevrolet', 'KIA', 'Mercedes’)  
        b = ('BMW’, 'Ford’)
        c = a + b  	    #  ('Chevrolet’, 'KIA’, 'Mercedes’, 'BMW’, 'Ford’)
        d = b * 2   	   #  d = ('BMW’, 'Ford’, ’BMW’, 'Ford’)


## 연습

* a = (1,2,3,4,5) 이고 b = (6,7,8,9) 이라고 할 때

  a+b를 하여 합해 c로 놓습니다.

  이어서 곱하기를 사용하여 ＂a’ 를 3번 반복해 d를 만들어봅니다.

  이를 각각 프린트 해봅니다.

* 이어서 Tupple의 아이템을 교체할 수 없다는 것을 확인해봅니다.

      b[0] = 10을 해서 교체하려고 하면 에러가 발생합니다.
       (물론 튜플에는 이런 메소드가 없습니다.)


##주의
* 튜플에 원소가 하나인 경우 integer로 오인할 수 있어 (5,) 식으로 표기함
 



##코딩 부분


In [0]:
a = (1,2,3,4,5)
b = (6,7,8,9)

c = a+b
print (c)

d = a*3
print (d)

(1, 2, 3, 4, 5, 6, 7, 8, 9)
(1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5)


위에서 아래와 같은 것을 하면 에러가 발생/ 교체가 안 되기 때문
b[0] = 10    # 에러가 발생



##추가 연습 

* 인덱싱 및 자르기를 해봅니다. 

  	    a = (‘Chevrolet’, ‘KIA’, ‘Mercedes’)  
  	    a0 = a[0] 	# 0의 위치에 ‘Chevrolet’ 이 있음을 알게 합니다.
  	    a1 = a[1:2]   # 1 위치 전에 자르고 2(즉 3번째) 이전에 자른다. 
  	    a2 =a[:2] 	# 이는 2 (즉 3번째 위치) 이후 것은 모두 자른다.

* 다음과 같은 method도 있습니다.

        count()    # tupple에서 특정 value가 몇 번이나 나오는지 반환합니다. 
        index()    # tupple에서 특정 value가 어디에 있는지를 반환한다

* 이를 연습해봅니다.


In [0]:
a = ('Chevrolet', 'KIA', 'Mercedes')  
print (a[0]) 
print (a[1:2]) 
print (a[:2]) 
print (a.count('KIA'))
print (a.index('Mercedes'))

Chevrolet
('KIA',)
('Chevrolet', 'KIA')
1
2


##결과 콘솔

        Chevrolet                 # 0의 위치에 무엇이 있나 확인
        ('KIA',)                      # 1위치 전에 자르고 2위치부터 자르면 1만 남음 
        ('Chevrolet', 'KIA')      # 2위치 이후 것은 자른다
        1                              # KIA가 1번 나옴을 확인
        2                              # Mercedes의 위치 확인

#Lesson 19 – Set


##학습 내용
* Set은 dictionary처럼 { }으로 표현됩니다. set 은 아이템을 변경할 수 있으나 반복되는 아이템이 있으면 버립니다.


##문법

* Set 은 다음과 같은 형태를 취합니다.

         myset = { 아이템1, 아이템2, 아이템3, 아이템4}


 
##설명
* 예를 들어 리스트에서 아래와 같이 ‘John’ 이 반복되어도 상관 없으나

        mylist = [‘John’, ‘Tom, ‘John’, ‘Josh’, ‘Peter’]

* 이를 set 으로 바꾸면 반복되는 ‘John’ 이 사라지고 				

        {‘John’, ‘Tom, ‘Josh’, ‘Peter’} 만 남게 됩니다.


        myset = set(mylist) 
        # {‘John’, ‘Tom,  ‘Josh’, ‘Peter’}

* Set 은 수학의 집합처럼, a, b 가 set 이라면 합집합 교집합 및 여집합을 만들 수도 있습니다.  

        c = a.intersection(b)   	# c = a & b 교집합
        d = a.union(b)              # d = a | b 합집합
        e = a.difference(b) 		# e = a - b  여집합

## 연습

* 아래와 같이 두 set 이 있을 때

        a = {1,2,3,4,5} 
        b = {5,6,7,8,9}

* 다음 set을 만들어서 프린트 해봅니다.

        c = a.intersection(b)   	# c = a & b 교집합
        d = a.union(b)              # d = a | b 합집합
        e = a.difference(b) 		# e = a - b  여집합



##주의
* Set 은 중복이 안되기 때문에 겹쳐지는 아이템이 없어짐을 주목합니다.



##코딩 부분

In [0]:
a = {1,2,3,4,5} 
b = {5,6,7,8,9}
c = a.intersection(b)   
d = a.union(b)
e = a.difference(b) 

print (c)
print (d)
print (e)

    
##추가 연습 

* 메소드 ‘add’ 는 아이템 ‘하나’ 를 더 하게 해줍니다.


        set이름.add(아이템)


* 메소드 ‘update’ 는 ‘여러’ 개의 아이템을 더하게 해줍니다.


        set이름.update(아이템1, 아이템2)


* 메소드 ‘remove’ 는 하나를 제거 해줍니다.


        set이름.add(아이템)

* Set a 에게 아이템 {1,2,3,4,5} 을 주고 
  	
6 을 add 하고

7,8 을 update 하고

5를 remove 하며 각 과정마다 a 를 프린트하여 확인해 봅니다.



In [0]:
a = {1,2,3,4,5} 

a.add(6)
print (a)

a.update({7,8})
print (a)

a.remove(5)
print (a)

{1, 2, 3, 4, 5, 6}
{1, 2, 3, 4, 5, 6, 7, 8}
{1, 2, 3, 4, 6, 7, 8}


#Lesson 20 - Making Functions


##학습 내용
* 함수는 반복해서 재사용 할 수 있는 '수학적 공식' 같은 것이라 생각하면 됩니다. 그럼 함수를 한 번 만들어 보겠습니다. 


##문법

            def 함수이름():         # 함수 선언
		        함수기능            # 함수 기능 주기
        
            함수이름                # 함수 호출

 
##설명
* ‘brake’란 함수를 만듭니다. 우리는 브레이크의 기능을 알고 있습니다. 밟으면 차가 선다는 기능이지요.  def 문으로 이 함수를 선언합니다.

        def brake():     

* 아직은 초보적인 기능 밖에 모르므로 print 문으로 기능을 만들어 봅니다.  

        print (‘Hit the brake, then the brake will stop the car.’)

* 그리고 마지막으로 함수를 call 하여 함수가 작동하도록 해야 합니다. 

        brake()     

* 즉 함수를 선언하여 기능을 정의하고 외부에서 그 함수를 호출하는 방식입니다.



## 연습

* 비슷하게 하나를 더 해 봅니다.

    	# wiper란 function을 define하고 함수 기능을 준다.	
    	def wiper():		
    		print ('When it is rainy, it is useful.') 	
    	
    	wiper()			# 호출해 봅니다.



##주의
* def 함수이름():  에서 콜론을 빠트리지 않도록 주의합니다.



##코딩 부분

In [0]:
def wiper():
    print ('When it is rainy, it is useful.')

wiper()

When it is rainy, it is useful.


##추가 연습 

* x의 세제곱을 구하는 함수를 만들어 보겠습니다.
* 함수이름은  mycal()로 합니다.
* x = 3이라고 정의하고
* x의 세제곱을 출력하도록 함수를 만든  후
* mycal()을 호출합니다.


In [0]:

def mycal():
	x = 3
	print (x**3)

mycal()


27


#Lesson 21  -  Parameters and Arguments


##학습 내용
* Parameter(매개변수)는 변수로 function의 input 역할을 하는 것입니다. 
* 그리고 Argument는 파라미터의 값으로 함수에서 사용되는 입력되는 값을 의미합니다.


##문법

         def 함수이름(파라미터1, 파라미터2):    # 함수 선언
            함수기능                           # 함수 기능 주기
        
        함수이름(아규먼트1, 아규먼트2)          # 함수 호출


 
##설명
* 수학의 함수에서 f(x) = 10x라고 정의했다고 가정하면, 여기서 x가 매개변수인 파라미터이고 x에 3을 넣어주면 답은 30이 되는데 여기서 3이 x의 값에 해당되는 아규먼트라고 이해를 하면 됩니다.  
* 우리는 앞에서 ‘brake’라는 아주 간단한 함수를 만든 적 있습니다. 여기서는 * parameters & arguments를 사용하여 function ‘brake’ 를 다시 만들어 봅니다. 
* 우선 parameter를 함수 brake에 주어봅니다. 

        def brake(pedal,pad):

* 그리고 pedal과 pad에 다음과 같은 아규먼트 값을 줍니다.

		pedal의 argument값은 "Hit the brake,"
		pad의 argument값은  " then the brake will stop the car."


## 연습
* 앞의 brake함수를 완성해 보겠습니다.

        # 파라미터를 주고 함수를 정의합니다.
		
		def brake(pedal,pad):
			print(pedal + pad)

	    # 아래에서 argument의 값을 주고 호출합니다.
		
		brake("Hit the brake,"," then the brake will stop the car.")


##주의
* 아규먼트의 개수가 파라미터와 일치해야 합니다.




##코딩 부분


In [0]:
def brake(pedal,pad):

   print (pedal + pad)

brake("Hit the brake,"," then the brake will stop the car.")

Hit the brake, then the brake will stop the car.


##추가 연습 

* 함수 my_function3를 선언합니다
* parameter로 (wish, country) 두 개를 부여합니다.
* 두 개를 합하여 아래를 프린트하게 아규먼트를 부여합니다

        I would like to go to Korea.
        I would like to visit Spain.
        I would like to live in Australia.



In [0]:
def my_function3(wish, country):
   	print (wish + country)

my_function3("I would like to go to", " Korea.")
my_function3("I would like to visit", " Spain.")
my_function3("I would like to live in", " Australia.")

I would like to go to Korea.
I would like to visit Spain.
I would like to live in Australia.


#Lesson 22  - Conditional Statements in Functions


##학습 내용
* 조건문이 들어가는 function을 만들어 보겠습니다. if – elif - else 문을 함수 안에 넣어 보겠습니다.


##문법
* 다음과 같은 형태로 함수 안에 조건문을 넣은 함수를 만들어 봅니다.

        def 함수이름(parameters):
		    if condition:
			    return
		    elif condition:
			    return
		    else:
			    return
      	
        함수이름(arguments)


 
##설명
* 함수 안에 기능을 넣을 때 조건문을 넣어서 만들 수 있습니다.
* def 함수이름(): 과 끝에 호출 하는 중간에 일반적인 조건문과 같은 방법으로 조건문을 넣는다고 생각하면 됩니다.


## 연습
* 우선 if –else로 엑셀러레이터라는 함수를 만들어 봅니다. 

        Function : accelerator
        Condition : speed 가 시속 80km 를 넘으면 "Slow down!" 을 출력합니다.
        else; 그 외의 경우는 Go!를 출력한다


##주의
* return은 결과 값을 반환하는 것을 의미합니다.



##코딩 부분

In [0]:
def accelerator(speed):
	if speed > 80:
		print ("Slow down!")
	else:
		print ("Go!")

accelerator(85)

Slow down!



##추가 연습 

* if - elif - else 로 만들어 봅니다.

		x = 5
		y = 10
	    이라고 하고
		함수 cal() 을 선언하고
			if x == y 면 print (‘They are the same.’)
			elif x > y 이면 print (‘x is larger than y’)
			else: print (‘y is larger than x.’)
		로 하고 함수 cal() 을 호출함으로 끝나는 함수를 만들어 봅니다.



In [0]:
x = 5
y = 10

def cal():
	if x == y:
		print ("They are the same.")
	elif x > y:
		print ("x is larger than y")
	else:
		print ("y is larger than x.")

cal()

y is larger than x.



#Lesson 23 -  Importing Modules


##학습 내용
* import 는 이미 만들어진 모듈을 호출하여 사용할 수 있게 해주는 역할을 합니다. 우리는 함수에 대해 배웠습니다.  이미 만들어진 모듈을 호출하는 것을 배워봅니다.


##문법
* 다음과 random 이라는 모듈은 값을 무작위로 제공 할 수 있는 모듈입니다. random 내에서 정수를 임의로 선택하는 함수인 randint 를 가져오면 무작위로 정수 값을 얻을 수 있습니다.

         from random import randint

* 시간과 날짜와 관계된 프로그램을 수행하고 싶다면 최선의 방법은 datetime 모듈을 import 할 수 있습니다.


         from datetime import datetime

##설명
* randint 는 특정 영역 안에 있는 정수를 임의로 선택하게 해줍니다. 


         toss = randint(0, 1) # 0과 1 둘 중 하나를 토스 값으로 줍니다.


* datetime 은 현재시간의 값을 줍니다. 


        current_time = datetime.now()   	# 현재 시간 값
        c_year = current_time.year      	# 현재 시간에서 연도 값
        c_month = current_time.month  	  # 현재 시간에서 월 값
        c_day = current_time.day       	 # 현재 시간에서 일 값


## 연습

* 주사위를 던져 1에서 6 사이 임의로 나오는 값을 구해봅니다.
* 변수 current_time = datetime.now() 를 만들어 봅니다.
* 변수 c_year = current_time.year 를 만들어 봅니다.
* 변수 c_month = current_time.month 를 만들어 봅니다.
* 변수 c_day = current_time.day를 만들어 봅니다.
* 이를 출력해봅니다.


##주의
* datetime 외에도 time 이란 아주 살짝 다른 모듈이 있습니다.




##코딩 부분

In [0]:
from random import randint

dice_toss = randint(1, 6)
print (dice_toss)

from datetime import datetime

current_time = datetime.now()
c_year = current_time.year 
c_month = current_time.month
c_day = current_time.day
print (c_year)
print (c_month)
print (c_day)


3
2019
12
2


  
##추가 연습 

* ‘math’ module 을 사용해 보겠습니다.
* 우선 math module 을 import 해야 합니다
* 다음 4개의 함수를 배워 봅니다: ceil, floor, max, min

        ceil()      # '올림' 의 기능을 합니다. 주어진 숫자보다 같거나 더 큰 정수를 결과 값으로 받습니다.
        floor()     # '내림' 의 기능을 합니다. 주어진 숫자보다 같거나 더 작은 정수를 결과 값으로 받습니다.
        max()       # 주어진 숫자 중에서 가장 큰 값을 보여 줍니다. string 일 경우 알파벳에서 가장 늦은 순서의 알파벳을 보여 줍니다.
        min()       # 주어진 숫자 중에서 가장 작은 값을 보여 줍니다. string 일 경우 알파벳에서 가장 빠른 순서의 알파벳을 보여 줍니다.


In [0]:
import math               

print (math.ceil(0.5))           
print (math.floor(0.5))  
print (max([1, 2, 3]))
print (max('Chevrolet'))
print (min([1, 2, 3]))
print (min('Chevrolet'))

1
0
3
v
1
C


#Lesson 24      -      For Loop


##학습 내용
* 파이썬에는 for loop 과 while loop이 있습니다. for는 어떤 법위내에서 한정된 루프를 돌릴 때 사용합니다. 


##문법
* for 는 다음과 같은 형태를 취합니다.

         for 변수 in 대상:
		        수행 내용


 
##설명
* 다음과 같이 리스트의 범위 내에서 숫자를 출력하도록 할 수 있습니다.

        a=[1,2,3,4,5]

        for var in a:
            print (var) 	# 이 다섯 개 아이템들을 차례로 출력됩니다.

* string 도 마찬가지로 할 수 있습니다.


        b = [＂KIA＂, ＂BMW＂}

        for s in b:
		    print (s)



## 연습

* 스트링＂car＂를 for i in range(5): 를 사용하여 다섯 번 출력해 봅니다. 
* 숫자 1 부터 10 까지 더해봅니다. 
* ‘KOREA’ 를 알파벳으로 하나씩 출력해 봅니다.



##주의
* range 를 쓸 때  range(2,5) 라고 하면 2 부터 4 (5가 아님) 가 적용됨
* range(5)는 0,1,2,3,4 즉 다섯 번이 적용됨




##코딩 부분

In [0]:
for i in range(5):
	print ('car')

total = 0
n = 10      
for i in range(1, n + 1): 
	total += i     
print (total)

for character in 'KOREA':
	print (character)


car
car
car
car
car
55
K
O
R
E
A


  
    
##추가 연습 

* range 로 세 가지를 더 해 봅니다.

(1) 2부터 4까지 숫자를 제곱과 세제곱을 만들어 출력합니다.

    끝에 ‘That’s it’ 을 출력합니다

(2) range에서 시작 값과 마지막 값 그리고 각 단계별로 더하거나 빼는 값을 넣는 아래 문법을 사용하여 10에서 0까지 레인지에서 3씩 빼 나가는 것을 해봅니다.

        range(start_value, end_value, step)
    
(3) 마찬가지로 10에서 20까지 3 씩 더해 가는 것도 해봅니다.



In [0]:
for i in range(2, 5):
	print (i**2, i**3)
print ("That's it")

for n in range(10, 0, -3):
	print (n)

for n in range(10, 20, 3):
	print (n)

4 8
9 27
16 64
That's it
10
7
4
1
10
13
16
19


#Lesson 25     While / While - Else Loop


##학습 내용
* "while loop" 은 조건이 False 가 될 때까지 반복할 때 쓰이는 루프입니다. 


##문법
* while 은 다음과 같은 형태를 취합니다.

         while condition:
		    body


 
##설명
* while 은 우리가 for 처럼 몇 회, 숫자 몇 까지 같이 명확하게 주어지는 것이 아니라 언제 끝날지 확실히 모를 때 또는 몇 번이나 반복할지 확실하지 않을 때 (조건이 될 때 까지) 주로 사용됩니다. 
* 예를 보겠습니다. 아래는 정수가 5이하일 때 세제곱을 프린트 하라고 하는 것입니다. (물론 여기서는 간단한 것이니 뻔히 몇 회 루프가 돌지 보입니다) 

        i = 1
        while i <= 5:
		    print (i ** 3)
		    i += 1       		# 이는 i = i + 1 과 같은 표현임 


## 연습

*앞의 예를 다른 숫자로 실행해 봅니다.

		i = 15 에서 시작하여
		1 > 0 보다 큰 조건에서 
			i 의 제곱을 출력해 봅니다.
			이어서 i 에서 2씩 빼 나갑니다.



##주의
* i -= 1 은 i = i – 1 과 같은 뜻입니다.




##코딩 부분

In [0]:
i = 15

while i > 0:
    print (i ** 2)
    i -= 2

225
169
121
81
49
25
9
1


##추가 연습 

* While-else
	
이번에는 조건문의 else 를 끝에 추가해 문장을 하나 더 출력해 봅니다.

        i = 1

        while i <= 3:
            print (i)
            i += 1
        else:
            print ("That's it")

* 이를 실제로 실행해 봅니다.


In [0]:
i = 1

while i <= 3:
	print (i)
	i += 1
else:
    print ("That's it.")

1
2
3
That's it.


#Lesson 26    -   Break/Continue


##학습 내용
* for 나 while loop 에서 빠져나오려면 break 문을 사용합니다. loop 를 빠져나오지 않고 해당 조건만 건너뛰고 싶을 때는 continue 문을 사용합니다. 


##설명
####break
* break 문을 배워 봅니다. 이는 특정 조건이 맞을 때 루프를 중단하라고 명령하는 것입니다. (아래에서는 4가 되면 더 이상 루프를 돌리지 말라는 break 문을 넣어 본 것입니다.) 

         	i = 1

	        while i <= 10:
		        print (i)
		        if i == 4:
			        break
		        i += 1

####continue
* continue 문을 배워 봅니다. 이는 현재 진행 중인 루프를 거부하고  마지막 조건을 계속 다시 위로 보내는 역할을 합니다.

         	i = 1

	        while i <= 5:
		        i += 1
		        if i == 4:
			        continue
	
        	print (i)


## 연습
* 1에서 5까지 출력을 하는데 중간에 3이 오면 break 를 하게 해봅니다.
* 1에서 5까지 출력을 하는데 중간에 3이 오면 continue 를 하게하여 뛰어 넘게 해봅니다.



##주의
* 어떤 범위까지 출력을 할 때 그 범위에 대한 정확한 설정이 중요합니다. 




##코딩 부분

In [25]:

i = 0

while i <= 5:
	print (i)
	if i == 3:
		break
	i += 1

i = 0

while i <= 4:
	i += 1
	if i == 3:
		continue
	print (i)

0
1
2
3
1
2
4
5


##추가 연습 

(1) 	짝수 출력
* 홀수를 건너 띄게 하여 짝수만 1 부터 10 사이에서 출력하게 해봅니다.
* 이를 위해서는 나머지가 0 이 되지 않는 숫자는 continue 하게 만들어야 합니다.

(2) 	홀수 출력
* 마찬가지로 짝수를 건너 띄게 하여 홀수만 1 부터 10사 이에서 출력하게 해봅니다.
* 이를 위해서는 나머지가 1이 되지 않는 숫자는 continue 하게 만들어야 합니다.



In [0]:
for num in range(1, 11):
	if num % 2 == 0:
		print (num)
	continue

for num in range(1, 11):
	if num % 2 == 1:
		print (num)
	continue

2
4
6
8
10
1
3
5
7
9


#TASK 27     -        File I/O


##학습 내용
* 에디터를 벗어난 파일을 다루기 위해 "File I/O" 를 학습하여야 합니다. 이는 추후 학습에서 꼭 필요한 과정으로 사무 자동화 등에 유용하게 쓰이는 부분의 기초적인 내용입니다. 


##문법
* 3개의 아이템을 가진 test list 를 만들어 봅니다. 이어서 f(=file) 라는 variable 으로 test.txt 를 여는데 "w"(=writing) 모드로 엽니다. 이 경우 여러분이 보시는 main.py 우측에 test.txt 가 열림을 볼 수 있습니다. 이어서 close 를 하여 마무리 합니다. 이 과정이 리스트를 통해 파일을 만들고 이를 쓰기 모드에서 열어보았다가 닫는 과정입니다. 

        test_list = ["KIA", "Porsche", "Maserati"]
        f = open("test.txt", "w")
        f.close()


 
##설명
* 하지만 아직 리스트에 있는 아이템이 test.txt 파일에 쓰여 진 것은 아닙니다. 따라서 for loop를 활용하여 리스트에 있는 string item들이 아래와 같은 방법을 통해 파일에 입력 되도록 합니다.

        for value in test_list:
		    f.write(str(value))


## 연습
* 앞에 경우에서 리스트를 출력해보면 이 경우 알파벳 순서대로 
	
        "KIAMaseratiPorsche" 
	
이런 형식으로 출력이 됩니다. 

빈공간이 없으므로 붙어서 나오게 되지요. 

\n 을 넣어서 줄 바꾸기를 하여 세로로 나오도록 해봅니다.

내용을 그냥 숫자 1,2,3,5,6 으로 된 리스트로 만들어 보겠습니다.

 

##주의
* \n 을 넣어서 줄 바꾸기를 하여 세로로 나오도록 할 수 있습니다.



##코딩 부분

In [0]:
test_list = [ 1, 2, 3, 4, 5, 6]
f = open("test_list.txt", "w")

for item in test_list:
	f.write(str(item) + "\n")
	print (str(item))

f.close()

1
2
3
4
5
6


##추가 연습 

* 이제 읽기를 해봅니다.
* 파일 읽기는 "r" 모드에서 되며 우리의 입장에서는 제대로 읽었는지 알 수가 없으므로 이를 프린트 함으로 확인합니다.
* 앞의 페이지와 같은 방법으로 파일을 만든 후 그 다음에 읽기를 해봅니다. 
* 그런데 쓰기 모드에서 새로운 쓰기 모드를 하면 앞의 리스트를 덮어 써 버립니다.
* 그래서 앞에 것을 지우지 않게 추가만 할 경우 append 모드 "a" 로 씁니다.
* 4를 “w” 모드로 쓰면 1,2,3 은 없어지고 4 만 남습니다.
* 하지만 “a” 모드로 쓰면 1,2,3 에 4 가 더 해집니다.
* 연습을 해봅니다.


In [0]:
test_list = [1,2,3]
f = open("test_list.txt","w")

for item in test_list:
	f.write(str(item) + "\n")

fr =  open("test_list.txt","r")
print (fr.read())

frp = open("test_list.txt","a")
frp.write('4')

frp.close()


    # main.py에서는 1,2,3 만 보이나 test_list.txt 를 보면 1,2,3,4 가 나온다. 

###Extra Practice

(1)  Colab에서 자기 PC에 있는 파일 불러 오기

* PC에 있는 파일을 여기 작업 공간으로 불러온다.

In [26]:
from google.colab import files
uploaded = files.upload()



#Lesson 28    -    Making a Class 


##학습 내용
* 함수, 변수, data 등을 포함한 큰 오브젝트 class가 있습니다. 한 번 클래스를 만들어 보겠습니다. 여기서 self 및 _init_의 개념도 공부합니다. 클래스는 다른 모듈들과 구분하기 위해 대문자로 표기합니다. 


##설명
* 우선 빈 클래스를 만들어 봅니다.

         class Car():
		    pass

* 위의 클래스는 빈 클래스 입니다. 여기에 요소들을 붙여 봅니다.
* 우선 클래스 "Car" 를 만들고 이 안에 object 들이 있음을 알립니다.
* 그리고 initializer 라 불리우는 _init_을 넣습니다. 
* _init_ 은 initializer, 즉 초기에 뭔가 정해주는 성질을 가집니다.
* 우리는 메소드를 만들 때  _init_() 을 사용하는데, 이 것의 첫 argument는  "self" 로 시작합니다.  

	    class Car(object): 
		    def __init__(self):

* 우리는 "self" 를 최초의 인자로 사용했습니다. 이는 다음에 오는 다른 argument 모두에게 "self" 를 부여하는 효과가 있습니다.  이는  마치 여기서 만들어지는 매소드는 이 클래스에 속한 것이라 선언하는 효과가 있습니다.

-----

* 우선 이해를 쉽게 하기 위해 _init_ 없이 클래스를 만들어 봅니다.

    	class Car(): 
	    	def brake():
		    	print ("Hit the brake!")

		    def pedal(self):
			    print ("Stop the car!")


* 위의 경우를 보면 함수 brake() 에는 “self” 가 없습니다. 
* 위를 실행해 보면 f.brake() 는 에러가 발생하나 f.pedal() 은 “Stop the car!”를 프린트 합니다. 왜 그럴까요?  
* 이는 brake가 아무런 argument 없이 시작되었는데 아래에서 아규먼트를 받기 때문입니다. 이는 자체적으로 “self” 를 통해서 이 클래스 내에서 쓰일 인자를 받아야 아래에서 이를 출력할 수 있다는 말입니다. (물론 self 말고 다른 (예를 들어 a 라는) 인자를 주어도 출력이 가능합니다. 하지만 self 를 주어야 그 다음에 올 인자들에게 self 를 부여할 수 있게 됩니다.) 
* 위의 경우에는 Hit the brake! 와 Stop the car! 를 출력합니다.


    	class Car(): 
    		def brake(self):
    			print ("Hit the brake!")

    		def pedal(self):
    			print ("Stop the car!")

    	f=Car()
    	f.brake()
    	f.pedal()


## 연습

* 이제 예로 차계부 Class 를 만들어 보겠습니다. 

(1) class Manager: 를 만듭니다. 

        class Manager():

(2) def init(self, model, color, year, mileage) 로 정의 합니다.

        def_init_(self, model, color, year, mileage):

(3) 이어서 self 로 정의된 변수를 변수랑 연결해 줍니다. 

        self.model = model
        self.color = color
        self. year =  year
        self.mileage = mileage

(4) 이들을 출력할 수 있게 argument 를 줍니다.

        mycar = Manager("Model: Camaro", "Color: Red", "Year: 2020", "Mileage: 12,345")


(5) 출력합니다.

        print (mycar.model)
        print (mycar.color)
        print (mycar.year)
        print (mycar.mileage)





##코딩 부분

In [0]:
class Manager(object):
    def __init__(self, model, color, year, mileage): 
       self.model = model
       self.color = color
       self.year = year
       self.mileage = mileage

mycar = Manager("Model: Camaro", "Color: Red", "Year: 2020", "Mileage: 12,345")

print (mycar.model)
print (mycar.color)
print (mycar.year)
print (mycar.mileage)


Model: Camaro
Color: Red
Year: 2020
Mileage: 12,345


####Video
[Lesson 28 - Lesson 38까지의 영상](https://youtu.be/zVIQuRShh9U)

#Lesson 29      -      Instance 



##학습 내용
* instance 는 클래스에 의해 생성되는 오브젝트로 각 인스턴스는 고유의 값을 가집니다. 


##문법
* “Car” 가 당신의 클래스라면 “my_car” 는 고유의 instance 입니다.
* “야구선수”가 당신의 클래스라면 “류현진”은 고유의 인스턴스가 될 것입니다.
* 이와 같이, 인스턴스는 클래스에 의해 생성되는 오브젝트로 각 인스턴스는 고유의 값을 가집니다. 


 
##설명
* 예를 들어 Car 라는 클래스를 만든다고 가정합니다.
* name(self) 이라는 메소드를 만든 다음
* 당신이 당신의 차에 ‘Bumblebee’ 란 이름을 주었다고 가정합니다
* 당신 차의 이름을 출력하는 것을 이 클래스의 함수로 주었다면
* 이 것이 인스턴스인 것입니다.


## 연습
* 연습을 만들어 봅니다.

        class Car(object):
		    def __init__(self, name):
			    self.name = name

	    mycar = Car("Bumblebee")
    	print (mycar.name)



##주의
* 클래스가 붕어빵 틀이라면 인스턴스는 틀에서 나오는 붕어빵이라고 생각하시면 이해가 좀 더 쉬울 것입니다.



##코딩 부분


In [0]:

class Car(object):
	def __init__(self, name):
		self.name = name

mycar = Car("Bumblebee")
print (mycar.name)

##추가 연습 

* 클래스 Car를 만듭니다.

		class Car(object):

모델, 컬러, 연도에 이름까지 차에 붙인다고 생각해 봅니다.

			def __init__(self, model, color, year, name):
				self.model = model
				self.color = color
				self.year = year
				self.name = name

제차 소개를 넣어 봅니다.

			def about_mine(self):
				return '제차는{}로 {}색 입니다. {}연도형으로 제가 이름을 {}라고  부르지요.'.format(self.model,self.color, self.year, self.name)

인자를 넣고 출력을 해봅니다.

    		mycar = Car('Palisade', '붉은', '2020', '코코')
    		print (mycar.about_mine()


In [0]:
class Car(object):
	def __init__(self,model,color,year,name):
		self.model = model
		self.color = color
		self.year = year
		self.name = name

	def about_mine(self):
		print ("제차는 {0}로  {1}색 입니다. {2}연도형으로 제가 이름을 {3}라고 부르지요.".format(self.model, self.color, self.year, self.name))


mycar = Car('Palisade', '붉은', '2020', '코코')
mycar.about_mine()


제차는 Palisade로  붉은색 입니다. 2020연도형으로 제가 이름을 코코라고 부르지요.


#Lesson 30  -   Creating and Accessing Methods 


##학습 내용
* 이제는 클래스 내에 메소드를 만들고 이를 사용하는 법을 배워 봅니다. 메소드는 클래스 내에 내장된 함수 같은 것입니다. 우리가 클래스를 만들 때 메소드도 같이 만들고 이를 사용해 보겠습니다. 


##설명
* 우선 Car 클래스를 만들어 봅니다.
* self 로 maker, color, style 의 오브젝트를 만듭니다.
* 브레이크 메소드는 다음을 프린트하는 기능을 가지도록 합니다.

         This is a {maker} with a good brake.

* 미러 메소드는 다음을 프린트하는 기능을 가지도록 합니다.

        This is {style} style with no blind spot.


* 와이퍼 메소드는 다음을 프린트하는 기능을 가지도록 합니다.

        It is {color} and comes with an excellent wiper.


* 아규먼트를 줍니다.

        mine = Car("Chevy", "dark blue", "cool")


* 호출하여 메소드를 사용합니다.

    	mine.brake()
    	mine.mirror()
    	mine.wiper()


## 연습

* Car 클래스를 만든 후
* brake method, mirror method, wiper method 를 만든 후 차례로 호출해 봅니다.



##코딩 부분

In [0]:
class Car(object):
	def __init__(self, maker, color, style):
		self.maker = maker
		self.color = color
		self.style = style

	def brake(self):
		print ("This is a {} with a good brake.".format(self.maker))

	def mirror(self):
		print ("This is {} style with no blind spot.".format(self.style))

	def wiper(self):
		print ("It is {} and comes with an excellent wiper.".format(self.color))


mine = Car("Chevy", "dark blue", "cool")

mine.brake()
mine.mirror()
mine.wiper()


This is a Chevy with a good brake.
This is cool style with no blind spot.
It is dark blue and comes with an excellent wiper.


##추가 연습 

* 3대의 차가 있다고 상상해 봅니다. 이 3대는 Chevrolet 의 2019 Corvette, Ford 의 2020 Mustang, Jaguar 의 2019 XF 라고 가정해 봅니다. 
* 셋 다 빠르다고 가정하여 is_fast = True 로 공통으로 주고 
* 이 차들의 현황을 프린트 할 수 있게 해봅니다. 
* 아규먼트는 다음과 같이 줍니다.

    	corvette = Car("Chevrolet", 2019)
    	mustang = Car("Ford", 2020)
    	xf= Car("Jaguar", 2019)



* 이를 프린트 해봅니다.

    	print (corvette.maker, corvette.year, corvette.is_fast)
    	print (mustang.maker, mustang.year, mustang.is_fast)
    	print (xf.maker, xf.year, xf.is_fast)

In [0]:
class Car(object):
	is_fast = True 

	def __init__(self, maker, year):
      		self.maker = maker
      		self.year = year

corvette = Car("Chevrolet", 2019)
mustang = Car("Ford", 2020)
xf = Car("Jaguar",2019)

print (corvette.maker, corvette.year, corvette.is_fast)
print (mustang.maker, mustang.year, mustang.is_fast)
print (xf.maker, xf.year, xf.is_fast)

Chevrolet 2019 True
Ford 2020 True
Jaguar 2019 True


#Lesson 31    -      Inheritance



##학습 내용
* 클래스 간의 상속(Inheritance)는 부모 클래스의 속성을 자식 클래스가 물려 받음을 의미합니다. 예를 들어 Car 라는 클래스가 있으면 "Car Class" 의 속성을 상속받는 자식 클래스인 Sedan, Truck 또는 SUV 와 같은 서브 클래스를 만들 수 있습니다. 해당 클래스에서 부모 클래스의 속성과 메서드를 사용할 수 있습니다.


##설명
* 아래의 예에서 간단히 어떻게 상속하는지를 보겠습니다. 자식 클래스를 만들 때 부모 클래스 이름을 괄호 안에 입력하면 됩니다.
* 그러면 따로 만들지 않아도 자식 클래스에는 그 속성이 전달됩니다.

    	class Car(object):       
    			body
    	class Truck(Car):


* 클래스 "Car" 를 만듭니다.
* Self로 "Category", "maker" 와 "year" 및 "color" 를 부모 클래스 "Car" 에 생성합니다.
* 서브클래스 괄호 안에 Car 를 넣어 "Truck(Car)" 클래스를 생성합니다.
* 서브 클래스에서 "Car Class" 의 메소드를 상속받았다고 가정하고 인자를 넣어서 프린트를 해봅니다.
* 이 것이 에러를 발생하는지 상속을 하는지를 확인 합니다.



##코딩 부분


In [0]:
class Car(object):
	def __init__(self, category, maker, year, color):
		self.category = category
		self.maker = maker
		self.year = year
		self.color = color

	def display_mycar(self):
		print ('MyCar: {}'.format(self.category))
		print ('Maker: {}'.format(self.maker))
		print ('Year: {}'.format(self.year))
		print ('Color: {}'.format(self.color))

class Truck(Car):
	pass

Truck_mine = Truck('Truck','Chevy', '2019', 'Red')
Truck_mine.display_mycar()

MyCar: Truck
Maker: Chevy
Year: 2019
Color: Red


##추가 연습 

* 클래스 “Car” 를 만듭니다.
* Self 로 “Category”, “maker” 와 “year” 및  “color” 를 부모 클래스 “Car” 에 생성합니다.
* 서브클래스 괄호 안에 Car 를 넣어 “Ecar(Car)” 전기차클래스를 생성합니다.
* 서브 클래스에서 Car Class 의 메소드를 상속받았다고 가정하고 인자를 넣어서 프린트를 해봅니다.
* 이 것이 에러를 발생하는지 상속을 하는지를 확인 합니다.


In [0]:
class Car(object):
	def __init__(self, category, maker, year, color):
		self.category = category
		self.maker = maker
		self.year = year
		self.color = color

	def display_mycar(self):
		print ('MyCar: {}'.format(self.category))
		print ('Maker: {}'.format(self.maker))
		print ('Year: {}'.format(self.year))
		print ('Color: {}'.format(self.color))

class Ecar(Car):
		pass

kona_mine = Ecar('eCar','HYUNDAI', '2019', 'Red')
kona_mine.display_mycar()

MyCar: eCar
Maker: HYUNDAI
Year: 2019
Color: Red


 ##추가 연습 

* 아래의 추가 적인 예를 보며 이해해 봅니다.


In [0]:
class Car():
	def __init__(self, engine, body):  
		self.engine = engine
		self.body = body

mycar = Car("2 lister engine", "beautiful shape.") 
print ("My car has {0} with {1}".format(mycar.engine, mycar.body))

class Truck(Car):
	pass

mytruck = Truck("5 liter engine", "tough body.")
print ("My car has {0} with {1}".format(mytruck.engine,mytruck.body))


My car has 2 lister engine with beautiful shape.
My car has 5 liter engine with tough body.


##Lesson 32 - Overriding


##학습 내용
* Method overriding 은 부모 클래스의 메소드를 자식 클래스에서 다시 정의 하는 것입니다. 예를 들어 부모 클래스의 특징에서 특정 메소드가 특정한 기능을 하는데 자식 클래스에서는 다른 것은 다 같으나 이 것만은 달라야 할 때가 있습니다. 이런 때 overriding 이 필요합니다.


##설명
* 아래의 예에서 간단히 보겠습니다. 
* Car 라는 부모 클래스의 메소드 중 하나인 톨비를 자식 클래스 Truck에서 다시 다른 값으로 정의 하는 것을 해 봅니다.
* 부모 클래스에서 메소드를 만들어서 세단의 톨비를 정해봅니다.

    	def display_car(self):
    		print ("The toll for a sedan is {} dollars". format(self.toll))


* 자식 클래스에선 이를 뒤엎는 메소드를 만들어서 트럭의 톨비를 정해봅니다. 

    	def display_car(self):
    		print ("The toll for a truck is {} dollars". format(self.toll))



* 아규먼트에서 각 각 다른 톨비를 정하고 호출합니다.

    	pay_toll = Truck("3")
    	pay_toll.display_car()
    	pay_toll = Car("2")
    	pay_toll.display_car()



*이제 단계별로 오버라이딩을 보겠습니다.

## 연습 1
* 아래에서는 Car 란 클래스를 만들고 gasoline 을 넣는다는 것으로 mother class 에 정해주고
* sub class Ecar 에서는 전기차 이므로 electricity 를 넣는다는 것으로 오버라이딩 해보겠습니다.
* 우선 mother class 만 있는 경우를 봅니다.

##코딩 부분


In [0]:
class Car(object):
	def __init__(self, category, fuel):
		self.category = category
		self.fuel = fuel

	def display_mycar(self):
		print ("My car is {0} and it requires {1} for fuel.". format(self.category,self.fuel))

mysportscar = Car("an old car","gasoline")
mysportscar.display_mycar()

My car is an old car and it requires gasoline for fuel.


## 연습 2

* 이어서 오버라이딩 하지 않고 inheritance 만 하는 경우를 보겠습니다




##코딩 부분

In [0]:
class Car(object):
	def __init__(self, category, fuel):
		self.category = category
		self.fuel = fuel

	def display_mycar(self):
		print ("My car is {0} and it requires {1} for fuel.". format(self.category,self.fuel))

class Ecar(Car):
	pass

myecar = Ecar("a new car","electricity") 
myecar.display_mycar()
mysportscar = Car("an old car", "gasoline")
mysportscar.display_mycar()

My car is a new car and it requires electricity for fuel.
My car is an old car and it requires gasoline for fuel.



## 연습 3

* 마지막으로 오버라이딩 해봅니다.





##코딩 부분

In [0]:
class Car(object):
	def __init__(self, category, fuel):
		self.category = category
		self.fuel = fuel

	def display_mycar(self):
		print ("My car is {0} and it requires {1} for fuel.". format(self.category,self.fuel))

class Ecar(Car):
	def display_mycar(self):
		print ("My car does not require fuel. and it only requires {}.". format(self.fuel))

myecar = Ecar("a new car","electricity") 
myecar.display_mycar()
mysportscar = Car("an old car", "gasoline")
mysportscar.display_mycar()

My car does not require fuel. and it only requires electricity.
My car is an old car and it requires gasoline for fuel.


#Lesson 33 - Super


##학습 내용
* 자식 클래스에 있는 것을 다시 부모 클래스의 메소드로 재정의 하려면 뭐가 필요할까요? 이때 필요한 것이 “super” 입니다. “super” 라는 단어는 자식 틀래스 안에서 부모 클래스의 요소들을 불러옵니다. 우리는 이를 4단계에 걸쳐 보겠습니다.  super는 super(서브클래스명, self).마더클래스함수(파라미터) 형식으로 표현됩니다.



## 연습 1
* mother class 만 있는 것으로 하고 세단이 km당 200원의 기름을 소모하는 것을 만들어 보겠습니다. 





##코딩 부분

In [0]:
class Car(object):
	def __init__(self, model):
		self.model = model
	
	def calculate_mileage(self, km):
		self.km = km
		return km * 200

road = Car('Sonata')
print (road.calculate_mileage(10))

2000



##결과 콘솔

        2000  		# 당연히 10*200 = 2000이 나옵니다. 


## 연습 2

* 이제 서브 클래스인 트럭 클래스에서 트럭이 그냥 km 당 200원을 inheritance 하는 것을 보겠습니다. 

##코딩 부분


In [0]:

class Car(object):
	def __init__(self, model):
		self.model = model

	def calculate_mileage(self, km):
		self.km = km
		return km * 200

class Truck(Car):
	pass


road = Truck('road')
print (road.calculate_mileage(10))

2000


##결과 콘솔

         2000  		# 상속했으니 당연히 다시 10*200 = 2000이 나옵니다. 

## 연습 3

* 이제 오버라이딩하는 것을 해봅니다. (트럭은 km당 300원 든다) 






##코딩 부분



In [0]:
class Car(object):
	def __init__(self, model):
		self.model = model

	def calculate_mileage(self, km):
		self.km = km
		return km * 200

class Truck(Car):
	def calculate_mileage(self, km):
		self.km = km
		return km * 300

road = Truck('road')
print (road.calculate_mileage(10))

3000


##결과 콘솔

        3000  		# 오버라이딩되어 10*300 = 3000이 나옵니다. 




## 연습 4

* 마지막으로 super 를 사용하여 고속도로에서는 트럭도 200원만 드는 것으로 바꿉니다. highway.milage 란 함수를 정의 하는데 return 값을 (Truck 클래스의 self로 되어 있는) calculate.mileaage(km) 를 super 하여 서브 클래스의 공식 대신 마더 클래스의 공식으로 바꾸는 기능의 함수를 생성하는 형태로 합니다. 






##코딩 부분

In [0]:
class Car(object):
	def __init__(self, model):
		self.model = model

	def calculate_mileage(self, km):
		self.km = km
		return km * 200

class Truck(Car):
	def calculate_mileage(self, km):
		self.km = km
		return km * 300

	def highway_mileage(self, km):
		return super(Truck, self).calculate_mileage(km)

highway = Truck('highway')
print (highway.highway_mileage(10))


2000


##결과 콘솔

         2000  		# super 로 인해 10*200 = 2000이 나옵니다. 


#Lesson 34      -      Lambda


##학습 내용
* 파이썬의 람다(lambda) 함수는 런타임에 생성해서 사용할 수 있는 익명 함수 입니다. 일반적인 함수와 같이 정의해두고 쓰는 것이 아니고 필요한 곳에서 간단한 기능을 즉시 사용하고 버릴 수 있는 함수입니다.



 
##설명
* 우리는 variable 과 function 의 차이를 압니다. 두개의 variable 을 사용하여 아래를 계산하면 답이 4가 나옵니다. 

      	a = 2        
      	b = 2
      	print (a ** b)


*Function 으로 이를 만들면 (1) 정의하고 (2) 공식을 넣고 (3) 함수를 호출합니다. 그리고 역시 같은 답 4를 얻습니다. Variable 보다는 좀 복잡합니다. 

      	def sample(a,b):
         		 print(a ** b)

      	sample(2,2)


* 그런데 lambda function 을 사용하면 마치 variable 만 사용한 듯 하면서 function 으로 답을 구할 수 있어서 훨씬 간단합니다. 

      	a = lambda x, y: x ** y
      	print (a(2,2))


* 그럼 람다로 다 하지 왜 귀찮게 function 을 만드느냐고 할 수도 있겠으나 lambda 는 임시로만 사용되는 한계가 있는 함수입니다.
 

## 연습 1

* x,y,z 를 더하는 람다함수를 만들고 거기에 1,2,3 의 값을 줍니다. 
* x,y,z 를 곱하는 람다함수를 만들고 거기에 1,2,3 의 값을 줍니다. 
* z = x + y 인데 x 에 10, y 에 20 을 넣고 계산해 봅니다. 



##코딩 부분

In [0]:
sample1 = lambda x,y,z: x+y+z
print (sample1(1,2,3))

sample2 = lambda x,y,z: x*y*z
print (sample2(1,2,3))

z = (lambda x,y: x+y) (10,20) 
print (z)

6
6
30


 ## 연습 2

* y = x ** 3 에 x 에 7 을 대입하여 답을 구해보세요. 
* z = x3 + y3 인데 x 에 3, y 에 4 을 넣고 해보세요. 
* 위에 것을 print (z(3,4)) 로 해서 하고 윗 줄에서 (3,4) 를 빼 보세요. 




##코딩 부분

In [0]:
y = lambda x: x**3
print (y(7))

z = (lambda x,y: x**3+y**3) (3,4) 
print (z)

z = (lambda x,y: x**3+y**3) 
print (z(3,4))


343
91
91



## 연습 3

* 앞의 과에서 만들어 보았던  함수와 비교해 보겠습니다.
* 조건문을 람다로 만들어 보겠습니다. 짝수와 홀수를 출력해 봅니다. 



##코딩 부분

In [0]:
# 앞 과에서 썼던 방법
for num in range(1, 3):
	if num % 2 == 0:
		print (num)
		continue
for num in range(1, 3):
	if num % 2 == 1:
		print (num)
		continue

# 람다 (두 가지 방법)
print ((lambda x, y: x if x % 2 == 0 else y)(1,2))
print ((lambda x, y: x if x % 2 == 1 else y)(1,2))

mylist = list(range(1, 3))
print (list(filter(lambda n:n % 2 == 0 , mylist)))
print (list(filter(lambda n:n % 2 == 1 , mylist)))


2
1
2
1
[2]
[1]


#Lesson 35   -    Lambda & Functions


##학습 내용
* 앞 과에서 lambda function 에 대한 기본을 배웠습니다. 여기서는 이미 알려진 람다 관련 function 들을 간단히 배워보겠습니다. 



##설명

(1) 	map() function 
* map() function 과 함께 사용하면 람다 함수 사용의 장점을 명확하게 볼 수 있습니다.
* map() function 은 두 개의 argument 를 필요로 합니다.
* 예를 들어 0 부터 4 까지의 세제곱을 새로운 리스트로 만들고 싶다면 아래와 같이 한 줄로 쉽게 만들 수 있습니다. 결과는 [0,1,8,27,64] 로 표현됩니다.

        list(map(lambda x: x ** 3, range(5)))


(2) filter() function. 
* filter()는 원래 list에서 "True" 값으로 나온 아이템으로 새 list 를 작성합니다. 
* 아래의 예를 보면 추가 줄을 입력하지 않고 새로운 list[4] 를 얻을 수 있습니다. 

        list(filter(lambda x: 3<x < 5, range(6)))

(3)	reduce() function
* reduce() function 은 반복되는 프로세스를 줄입니다.
* 예를 들어 1 부터 10 까지를 차례로 더하라고 하면 본래 loop 를 돌려 반복해야 하나 reduce 에서는 1,2,3,4,5,6,7,8, 9,10을 다 더하지 않고 1 에다 2 를 더하면 (쉽게 말해서 하나의 시범을 보이면) 나머지를 계속합니다.

        print reduce(lambda x,y: x+y, [1,2,3,4,5,6,7,8,9,10])





## 연습

연습 1
* map() function 을 사용하여 5 부터 9 까지의 정수의 제곱을 만드세요.

연습 2
* filter() function을  사용하여 0 부터 10 까지의 숫자 중 3 보다 작은 숫자의 리스트를 만드세요.

연습 3
* reduce() function 을 사용하여 3 부터 5 까지를 곱해 보세요   *reduce는 파이썬 3에서는 기본이 아니라 에러 발생 가능.

       print (reduce(lambda x,y: x*y, [3,4,5]))


##코딩 부분

In [30]:
print (list(map(lambda x: x ** 2, [5,6,7,8,9])))
print (list(filter(lambda x: x < 3, [1,2,3,4,5,6,7,8,9])))


[25, 36, 49, 64, 81]
[1, 2]




##결과 콘솔

        [25, 36, 49, 64, 81]
        [0, 1, 2]
        60

##추가 연습 

* y = x ** 3 에 x 를 1 부터 5 사이 정수 모두를 대입하여 구해 보세요.

	
* mylist =  [1, 2, 3]에서 

	2를 초과하는 숫자만 반환하게 filter 를 사용해 보세요.
	
* mylist =  [-2, -1, 1, 2] 에서 

	양수만 반환하게 filter 를 사용해 보세요. 


In [0]:
y = list(map(lambda x: x**3, (range(1,6))))
print (y)

mylist =  [1, 2, 3]
print (list(filter(lambda x: x > 2, mylist)))

mylist =  [-2, -1, 1, 2]
print (list(filter(lambda x: x > 0, mylist)))


[1, 8, 27, 64, 125]
[3]
[1, 2]



##결과 콘솔

        [1, 8, 27, 64, 125]
        [1, 8, 27]
        [3]
        [1,2]


#Lesson 36   -    List Comprehension


##학습 내용
* List comprehension 또는 LC는 리스트를 쉽게 생성하게 해줍니다.
기존 list 또는 반복 가능한 객체에서 새 list 를 만들 때 유용합니다. 

##문법
* LC는 다음과 같이 표현 됩니다. 

		newlist = [공식 for 요소 in 입력Sequence]
 

##설명

* 숫자 list 에서 어떤 조건을 사용하여 새 목록을 만들 수 있습니다.
* 다음과 같이 (5 까지 숫자에서 2 를 곱하여) 10 까지 숫자에서 짝수를 만들 수 있습니다. 

    	even_ten = [x * 2 for x in range(6)]
    	print (even_ten)



* 답은 [0, 2, 4, 6, 8, 10] 로 나옵니다.
* 마찬가지로 홀수도 만들 수 있습니다.

    	odd_ten = [x*2+1 for x in range(5)]
    	print (odd_ten)




* 답은 [1, 3, 5, 7, 9] 로 나옵니다.


## 연습


* 아래 a,b 를 합하여 새로운 리스트를 만들어 봅니다.

	set “a” 에 “BMW” “FORD” “CHEVY” 를 넣고 

	set “b” 에 “2019” “2020”을 넣어 봅니다.



##코딩 부분


In [0]:
a = ['BMW' ,'FORD', 'CHEVY']
b = [2019,2020]

comb_car = [(c, y) for c in a for y in b]

print (comb_car)

[('BMW', 2019), ('BMW', 2020), ('FORD', 2019), ('FORD', 2020), ('CHEVY', 2019), ('CHEVY', 2020)]


##추가 연습 1

* mylist = [1,2,3,"a","b",4.0,5.0,True, False] 에서 newlist1 을 만들어 봅니다. newlist1 는 mylist 에서 integer 만 뽑아서 만듭니다. 이를 위해서 type() 이라는 메소드를 배워 봅니다.
* 이는 데이터 타입이 이는 데이터 타입이 무엇인지를 파악하는 메소드 입니다. 같은 리스트에서 float 만 뽑아서 newlist2 를 만들어 봅니다.


In [0]:
mylist = [1,2,3,"a","b",4.0,5.0,True, False]
newlist1 = [i for i in mylist if type(i)==int]
print (newlist1)

mylist = [1,2,3,"a","b",4.0,5.0,True, False]
newlist2 = [i for i in mylist if type(i)==float]
print (newlist2)

[1, 2, 3]
[4.0, 5.0]


##추가 연습 2

* 같은 리스트에서 string 만 뽑아서 newlist3 를 만들어 봅니다.
* 같은 리스트에서 Boolean 만 뽑아서 newlist4 를 만들어 봅니다.


In [0]:
mylist = [1,2,3,"a","b",4.0,5.0,True, False]
newlist3 = [i for i in mylist if type(i)==str]
print (newlist3)

mylist = [1,2,3,"a","b",4.0,5.0,True, False]
newlist4 = [i for i in mylist if type(i)==bool]
print (newlist4)

['a', 'b']
[True, False]


#Lesson 37 -   Dictionary Comprehension


##학습 내용
* Dictionary Comprehension, 또는 DC 는 인풋 시퀀스에 따라 새로운 
dictionary 를 만들게 합니다. 


##문법
* 다음과 같은 문법 형식을 가집니다.

         newdict  = [key: value for key, value in input sequence]


 
##설명
* 출력식이 키/값 쌍이며 결과과 사전으로 표현된다는 점을 제외하면 Set Comprehension 과 매우 유사합니다.
* 여기서는 추가로 zip 함수에 대해 간단히 배우겠습니다.
* zip 함수는 둘 이상의 list 또는 string 을 사용하며 항목 수가 동일한 경우 결합시키는 것이 가능합니다.

        new_dict = {key: value for key, value in zip(dict1, dict2)}


## 연습

* 두 개의 list 를 결합시키는데 하나의 요소는 key 다른 리스트의 요소는 value 로 만들어 봅니다. 

        cars = ['BMW', 'FORD', 'CHEVY', 'KIA']
    	years = [2019, 2020, 2018, 2019]
    	# 아래에서는 input sequence 를 zip 함수로 만들어서 사용해 
    	봅니다.
    	year_dict = {key: value for key, value in zip(cars, years)}


##주의
* zip 함수는 두 딕쇼너리의 항목 수가 동일한 경우에만 결합시키는 것이 가능합니다.



##코딩 부분

In [0]:
cars = ['BMW', 'FORD', 'CHEVY', 'KIA']
years = [2019, 2020, 2018, 2019]

year_dict = {key: value for key, value in zip(cars, years)}

print (year_dict)


{'BMW': 2019, 'FORD': 2020, 'CHEVY': 2018, 'KIA': 2019}


##추가 연습 

* 농구선수와 포지션을 합한 딕쇼너리를 만들어 봅니다.
* players 리스트는 ['Curry', 'Harden', 'James', 'Davis', 'Jokic']이고
* 이들을 position 별로 연결해 봅니다. (포인트가드(PG), 슈팅가드(SG), 스몰포워드(SF), 파워포워드(PF), 센터(C))
* 포지션 리스트는 positions = ['PG', 'SG', 'SF', 'PF', 'C'] 입니다.


    	players = ['Curry', 'Harden', 'James', 'Davis', 'Jokic']
    	positions =   ['PG', 'SG', 'SF', 'PF', 'C'] 
    	mydict1 = {key: value for key, value in zip(players,positions)}
    	print (mydict1)

* 야구 투수들의 승패와 방어율을 나타내 봅니다..
* pitchers 리스트는 ['Kim', 'Lee', 'Park', 'Choi'] 이고 
* kim은 13승 3패 1.99 lee는 15승 4패 2.05 park 16승 8패 2.65 choi 18승 11패 2.45의 기록을 가지고 있습니다.
* 이를 하나의 dictionary 로 나타내 봅니다.



In [0]:
players = ['Curry', 'Harden', 'James', 'Davis', 'Jokic']
positions =   ['PG', 'SG', 'SF', 'PF', 'C'] 

mydict1 = {key: value for key, value in zip(players,positions)}
print (mydict1)

pitchers = ['Kim', 'Lee', 'Park', 'Choi']
kim = [13,3,1.99]
lee = [15,4,2.05]
park = [16,8,2.65]
choi = [18,11,2.45]

mydict2 = {key: value for key, value in zip(pitchers, [kim,lee,park,choi])}
print (mydict2)

{'Curry': 'PG', 'Harden': 'SG', 'James': 'SF', 'Davis': 'PF', 'Jokic': 'C'}
{'Kim': [13, 3, 1.99], 'Lee': [15, 4, 2.05], 'Park': [16, 8, 2.65], 'Choi': [18, 11, 2.45]}


####Extra Practice

(1) 학생들의 국어, 영어, 수학 점수를 넣어봅니다.

		students = ['Kim', 'Lee', 'Park', 'Choi']
		kim = ['국어 92', '영어 78', '수학 88']
		lee = ['국어 89', '영어 88', '수학 72']
		park = ['국어 72', '영어 98', '수학 78']
		choi = ['국어 82', '영어 87', '수학 91']

In [0]:
students = ['Kim', 'Lee', 'Park', 'Choi']
kim = ['국어 92', '영어 78', '수학 88']
lee = ['국어 89', '영어 88', '수학 72']
park = ['국어 72', '영어 98', '수학 78']
choi = ['국어 82', '영어 87', '수학 91']
mydict3 = {key: value for key, value in zip(students, [kim,lee,park,choi])}
print (mydict3)

{'Kim': ['국어 92', '영어 78', '수학 88'], 'Lee': ['국어 89', '영어 88', '수학 72'], 'Park': ['국어 72', '영어 98', '수학 78'], 'Choi': ['국어 82', '영어 87', '수학 91']}


#TASK 38 - Other Useful Imports


##학습 내용
* Generator expression 은 특별한 종류의 comprehension 입니다. 이는 한 번에 하나의 값만 반환하는 generator 입니다. 

##문법
* 답을 전부 안 주고 하나씩만 준다면 그 것이 왜 필요할까요?
* 우리가 list comprehension 에 익숙해질 무렵이면 우리가 모든 답을 한꺼번에 필요하지 않은 경우도 있다는 것을 알게 될 것입니다. 또 모든 답을 미리 구한다는 것은 (우리가 현재 배우는 수준에서는 리스트가 아주 짧지만)  과도하게 메모리를 할당한다는 것을 의미하기도 합니다.
* 예를 들어 다음과 같이 합을 구하는 코드는 메모리에 제곱의 답 전체를 요구합니다. 예) sum([x*x for x in range(10)])
* 하지만 generator expression 인 아래의 예는 더 작은 메모리를 요구하게 됩니다.  (여기서는 범위가 10이지만 엄청나게 큰 수가 될 수도 있겠지요.) 예) sum(x*x for x in range(10)) 차이점은 list 또는 set 같은 컬렉션의 반복에는 이미 모든 값이 있지만 generator 는 하나의 데이터만 생성해서 그렇습니다. 


 
##설명
* 우리는 몇 줄만 만들었으므로 우리 수준에는 별로 중요하지는 않지만 수천 줄의 코드가 있으면 완전히 다른 이야기입니다. 과다 메모리 사용은 프로그램 속도를 늦출 수 있는 요소입니다. 무제한 데이터를 가진 경우나 데이터가 큰 상태에서 프로그램을 부분적으로 처리해도 될 경우 generator 를 사용 할 수 있습니다. 


## 연습


* 아래 연습에서 코드를 실행해 보시면 print 를 하면 하나씩 답이 나와서 다음을 요구할 때 다시 프린트를 요청해야 합니다.

    	gen = (x**3 for x in range(6))


##주의
* range 를 넘어서면 에러가 발생합니다.



##코딩 부분

In [0]:
gen = (x**3 for x in range(6))

# 아래에서는 처음 답 하나만 나오고 전체 레인지의 답이 나오지 않습니다.
print (next(gen)) 
# 한 줄을 더 입력하면 다음 것이 나옵니다.
print (next(gen)) 
# 역시 한 줄을 더 입력하면 다음 것이 나옵니다.
print (next(gen)) 

0
1
8


##추가 연습 

* 앞에 연습이 계속 하나씩 너무 단순히 작업이 되었다면 이를 좀 더 간단히 해보겠습니다.  
    
    	gen = (x**3 for n in range(6))
    
    	# 5개 출력
    	for i in range(5) :
        		a = next(gen)
        		print (a)
    
    	# 하나 더 출력
    	for a in gen :
        		print(a)


In [0]:
gen = (x**3 for x in range(6))

# 5개 출력
for i in range(5) :
    	a = next(gen)
    	print (a)

# 하나 더 출력
for a in gen :
	print (a)

0
1
8
27
64
125


####Extra Practice

(1) 비슷한 연습을 해보겠습니다.
* 0 부터 9 까지 제곱을 만들어서 하나씩 꺼내 보겠습니다.
        
        gen  = (x**2 for x in range(10))
        
        print (gen)
        
        # 이어서 값이 계속 나오다가 에러가 나올 때까지 프린트를 계속 해봅니다.


#PART3 - NumPy 기본 이해 (20분)

# Lesson 39 - Array 만들기

#### NumPy : Numpy는 Python의 과학 컴퓨팅을 위한 핵심 라이브러리입니다. 고성능 다차원 배열 객체와 이러한 배열 작업을위한 도구를 제공합니다. 우리가 NumPy를 겉으로 눈에 띄게 많이 쓰는 것은 아니나 실제로 우리가 앞으로 쓸 도구가 NumPy 기반 위에서 움직이는 것이기 때문에 이에 대한 기본적인 이해는 필수라 할 수 있겠습니다.

####Arrays (배열)
* 우린 수학에서 행렬을 배웠을 것입니다.
* 여기서는 수학의 행렬과 다르게 배열(array)을 1차원 이상의 변수를 말한다 
*  1차원 리스트를 다차원으로 빠르게 쓴다고 생각하면 됩니다.
* 우리가 리스트를 쓸 때 바이너리 코드로 상당히 긴 것을 사용하는데 Numpy는 적게 써서 자원을 아껴 준다고 생각하면 됩니다. 
* 또한 메모리의 배열 적인 면에서도 더 빠르게 할 수 있는 요소를 가지고 있는데 이 것은 여기서 이야기 하지 않도록 합니다.
* 일단 예를 보면서 이해해 봅니다.

####Video
[NumPy 영상](https://youtu.be/zVIQuRShh9U)

####import
*우선 numpy를 import 합니다.

In [0]:
import numpy as np

##학습 내용
* NumPy를 배우는 것은 Array를 배우는 것과 거의 같다고 생각하면 됩니다.
* 우선 import를 하고, 1차원 array를 만들어 봅니다. 


##문법

        import numpy as np
        var = np.array([item1, item2.....])



## 연습

        import numpy as np
        var = np.array([item1, item2.....])
        print (var)

* 에디터에 넣고 출력해 봅니다.


##주의
* array는 수학의 행렬과 비슷하나 살짝 다른 점이 있다는 것을 염두에 두세요. 나중에 배웁니다.



##코딩 부분

In [0]:
import numpy as np
var = np.array([1,2,3])
print (var)

##학습 내용
* 2X3 array를 연습해 보겠습니다.
* 1차원이 

        1	2	3

이었다면

2차원이란

    	1	2	3
    	4	5	6

같은 모양을 말합니다

* 위의 경우 2 X 3 행렬인 것입니다. 

##문법

        import numpy as np
        var = np.array([[item11, item12.....], [item21, item22....]])



## 연습

        import numpy as np
        var = np.array([[item11, item12.....], [item21, item22....]])
        print (var)

* 에디터에 넣고 출력해 봅니다.


##주의
* [ [] ] 두 겹임을 기억하세요



##코딩 부분

In [0]:
import numpy as np
var = np.array([[1,2,3], [4,5,6]])
print (var)

##학습 내용
* 3차원 array를 연습해 보겠습니다.
* 이는 2차원을 겹으로 만드는 것으로 만듭니다.

##문법

        import numpy as np
        var = np.array([[item111, item112.....], [item121, item122....]], [item211, item112.....], [item221, item122....]]])



## 연습

        import numpy as np
        var = np.array([[item111, item112.....], [item121, item122....]], [item211, item112.....], [item221, item122....]]])
        print (var)

* 에디터에 넣고 출력해 봅니다.


##주의
* [ [[]] ] 세 겹임을 기억하세요



##코딩 부분

In [0]:
import numpy as np
var = np.array([[[1,2,3], [4,5,6]], [[1,2,3],[4,5,6]]])
print (var)

##Lesson 40 - Checking  Data Type, Dimension, Size, & Shape

##학습 내용
* dtype은 어떤 종류의 데이터인지 (integer인지 float인지 등)을 알려 줍니다.
* 일반적으로 float이 더 많은 메모리를 잡아 먹고, integer도 가능하면 메모리를 덜 쓰도록 해야지 결과가 빠르게 나오기 때문에 이는 중요합니다.



##문법

	variable.dtype
	



## 연습

	import numpy as np
	var = np.array([int1, int2.....])
	print (var.dtype)
	
 
	import numpy as np
	var = np.array([float1, float2.....] ) 
	print (var.dtype)



* 에디터에 넣고 출력해 봅니다.


* 개별 int32인지 int16인지 등의 사이즈 지정도 가능


##코딩 부분

In [0]:
import numpy as np
var = np.array([1,2,3])
print (var.dtype)

var = np.array([1.0,2.0,3.0])
print (var.dtype)

##학습 내용
* 자신의 array가 몇 차원인지 확인해 봅니다.


##문법

        variable.ndim




## 연습

        import numpy as np
        var = np.array([[item111, item112.....], [item121, item122....]], [item211, item112.....], [item221, item122....]]])
    	print (var.ndim) 	# n-dimension 확인/ 주피터 노트북 등을 쓰면 굳이 프린트 하지 않아도 되나 여기서는 프린트해서 확인


* 에디터에 넣고 출력해 봅니다.


##주의
* 


##코딩 부분

In [0]:
import numpy as np
var = np.array([[[1,2,3], [4,5,6]], [[1,2,3],[4,5,6]]])
print (var.ndim)


var = np.array([[[1,2,3], [4,5,6], [7,8,9]]])
print (var.ndim)


##학습 내용
* Shape은 자신의 행렬이 몇 X 몇 행렬인지를 보여 줍니다.


##문법

	        variable.shape



## 연습

        import numpy as np
        var = np.array([[item111, item112.....], [item121, item122....]], [item211, item112.....], [item221, item122....]]])
        print (var.shape)

* 에디터에 넣고 출력해 봅니다.



##코딩 부분

In [0]:
import numpy as np
var = np.array([[[1,2,3], [4,5,6]], [[1,2,3],[4,5,6]]])
print (var.shape)


var = np.array([[[1,2,3], [4,5,6], [7,8,9]]])
print (var.shape)

var = np.array([[1,2,3], [4,5,6]])
print (var.shape)


##학습 내용
* Size 는 얼마나 많은 메모리를 NumPy가 차지 하는지를 알려 줍니다.
* 주피터 노트북에서는 약간 다르게 할 수 있음 



##문법


        variable.size




## 연습

        import numpy as np
        var = np.array([[item111, item112.....], [item121, item122....]], [item211, item112.....], [item221, item122....]]])
        print (var.size)

* 에디터에 넣고 출력해 봅니다.


##주의
* 주피터 노트북에서는 개별 item size도 확인 가능


##코딩 부분

In [0]:
import numpy as np
var = np.array([[[1,2,3], [4,5,6]], [[1,2,3],[4,5,6]]])
print (var.size)

##Lesson 41 -	Indexing Items

##학습 내용
* 1차원 Array에서 아이템을 인덱싱 합니다.
* 리스트와 별 다름이 없습니다.
* 아시다시피 위치는 0,1,2,3,4,.....순서로 갑니다 

##문법

	variable[위치]
	



## 연습

	import numpy as np
	var = np.array([1,2,3,4,5])
	print (var[2])
	

##주의
* var[2]이므로 3번째 아이템이 출력이 되어야 합니다.



##코딩 부분

In [0]:
import numpy as np
var = np.array([1,2,3,4,5])
print (var[2])


##학습 내용
* 2차원 Array에서 아이템을 인덱싱 합니다.
* 행과 열의 위치를 인덱싱합니다. 예를 들어 2행의 3열이면 [1,2]로 인덱싱합니다.

##문법

	variable[행위치, 열위치]
	

## 연습

	import numpy as np
	var = np.array([[1,2,3,4,5], [6,7,8,9,10]])
	print (var[1,2])
	

##주의
* var[1,2]이므로 2행 3번째 아이템 8이 출력이 되어야 합니다.



##코딩 부분

In [32]:
import numpy as np
var = np.array([[1,2,3,4,5], [6,7,8,9,10]])
print (var[1,2])

8


##학습 내용
* 3X3 Array에서 아이템을 인덱싱 합니다.
* 행과 열의 위치를 인덱싱합니다. 예를 들어 2번째 2행의 3열이면 [0,1,2]로 인덱싱합니다.

##문법

	variable[행위치, 열위치]
	

## 연습

	import numpy as np
	var = np.array([[[1,2,3,4,5], [6,7,8,9,10], [11,12,13,14,15]]])
	print (var[0,2,3])
	

##주의
* var[0,2,3]이므로 3행 4번째 아이템 14가 출력이 되어야 합니다.
* 주피터 노트북에서는 약간 다르게 할 수 있음 (slicing)


##코딩 부분


In [31]:
import numpy as np

var = np.array([[[1,2,3,4,5], [6,7,8,9,10], [11,12,13,14,15]]])
print (var[0,2,3])

14


#### 1X3

    POSITION        0   1   2   
    ITEM            1   2   3   

#### 2X3

    POSITION        0,0 0,1 0,2 
    ITEM            1   2   3   
    POSITION        1,0 1,1 1,2 
    ITEM            6   7   8   
    
#### 3X3

    POSITION        0,0,0   0,0,1  0,0,2   
    ITEM            1       2       3      
    POSITION        0,1,0   0,1,1  0,1,2  
    ITEM            6       7       8     
    POSITION        0,2,0   0,2,1  0,2,2  
    ITEM            11      12      13     

##Lesson 42 -	Replacing Items


##학습 내용
* Array에서 특정 아이템을 변경합니다.


##문법

	variable[본래위치] = 바꿀내용
	

## 연습

	var = np.array([[1,2,3,4,5], [6,7,8,9,10]])
	var[1,1] = 100



##코딩 부분

In [0]:
import numpy as np
var = np.array([[1,2,3,4,5], [6,7,8,9,10]])
var[1,1] = 100
print (var)

##Lesson 43 -	Creating Arrays with 0s, 1s , & Others


##학습 내용
* 모두 0으로된 1로된 또는 다른 숫자로 된 Array를 만들어 본다.


##문법

	np.zeros(shape)
	np.ones(shape)
	np.full( (shape), 통일된 아이템)

## 연습

	# 0 3개로 된 1차원
	 np.zeros(3)
	# 1 2X3개로 된 2차원
	 np.oness([2,3])
	# 5 3X3 개로 된 3차원
	 np.full([3,3], 5)


##코딩 부분

In [0]:
import numpy as np
print (np.zeros(3))
print (np.zeros([2,3]))
print (np.zeros([3,3]))

print (np.ones(3))
print (np.ones([2,3]))
print (np.ones([3,3]))

print (np.full(3, 5))
print (np.full([2,3], 5))
print (np.full([3,3], 5))


##Lesson 44 - Creating Arrays with Random Numbers


##학습 내용
* 랜덤 인테저나 플롯으로 된 array를 생성해 본다.


##문법

	np.random.rand(shape)	        #float
	np.random.randint(shape)	     #integer

## 연습

	# 3X3 random float
	 np.random.rand(3,3)
	# 3X3 random integer
	 np.random.randint(최소숫자, 최대숫자-1, size =(3,3))

##주의
* 이 부분은 아래 에디터로는 되지 않아서 주피터 노트북에서 다음과 같이 한다

##코딩 부분

In [0]:
np.random.rand(3,3)

* Random Integer Arrays

In [34]:
np.random.randint(1,9, size=(3,3))

array([[1, 6, 5],
       [8, 6, 1],
       [6, 2, 5]])

##Lesson 45,	Methods for Creating Arrays


##학습 내용
* Identity Matrix 를 배웁니다.
* 반복하는 Matrix를 만들어 봅니다.

##문법

	np.identity(행렬사이즈) # 예를 들어 3이다 하면 3X3 행렬에 대각선으로 1,1,1 이 있는 매트릭스를 만듭니다.

## 연습

	np.identity(3)
	np.repeat(var, 차원)
	
##주의
* 아이덴티티 매트릭스는 행과 열의 숫자가 항상 같습니다. 2X2, 3X3, 4X4
* repeat은 주피터노트북에서 연습해 보겠습니다.

##코딩 부분

In [0]:
import numpy as np
print (np.identity(3))

In [0]:
var = np.array([[1,2,3]])
rep = np.repeat(var,3, axis=0)
print (rep)


##학습 내용
* np.copy() 메소드를 배웁니다.
* np.eye() 메소드를 배웁니다  (identity matrix와 같음) 
* np.empty() 메소드를 배웁니다.
* np.like() 메소드를 배웁니다.
* np.linspace() 메소들르 배웁니다.


##문법

	var.copy()
	np.eye(행렬수)
	np.empty(행,렬)
	
	np.zeros_like	# 기존 array의 차원을 본 뜬 0 행렬을 만듭니다.
	np.ones_like	# 기존 array의 차원을 본 뜬 1 행렬을 만듭니다.
	
	np.linspace(시작, 끝, 나누기갯수)	#시작부터 끝까지 사이를 나누기 갯수만큼 균등하게 나누어 줍니다.		


## 연습

	a = np.array([1,2,3])
	b = a.copy()
	print (b)

##주의
* 일부는 주피터노트북에서 연습해 보겠습니다.

##코딩 부분

In [35]:
import numpy as np


a = np.array([[1,2,3],[1,2,30]])
b = a.copy()
print (b)

print (np.empty(3,3))

print (np.linspace(0,10,4))


[[ 1  2  3]
 [ 1  2 30]]


TypeError: ignored

###내용

* np.eye() 메소드를 배웁니다  (identity matrix와 같음) 
* np.zeros_like	# 기존 array의 차원을 본 뜬 0 행렬을 만듭니다.
* np.ones_like	# 기존 array의 차원을 본 뜬 1 행렬을 만듭니다.
* np.arrange(시작점, 끝점, 단계)	# 예를 들어 0, 5, 1이라면 0부터 4까지 1씩 올라가면서 array 생성

In [0]:
a = np.eye(3)
print (a)
        
print (np.zeros_like(a))
print (np.ones_like(a))
print (np.arrange(0, 10, 2))